<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Utility Functions for the anglemania Package — anglemania_utils • anglemania</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Utility Functions for the anglemania Package — anglemania_utils"><meta property="og:description" content="A collection of utility functions used within the
anglemania package
for manipulating FBMs, calculating statistics, and selecting genes.
Replaces all NaN and Inf values in a numeric vector with NA."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">anglemania</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/anglemania_tutorial.html">anglemania Tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Utility Functions for the anglemania Package</h1>
    <small class="dont-index">Source: <a href="https://github.com/BIMSBbioinfo/anglemania/blob/master/R/anglemania_utils.R" class="external-link"><code>R/anglemania_utils.R</code></a></small>
    <div class="hidden name"><code>anglemania_utils.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A collection of utility functions used within the
<span class="pkg">anglemania</span> package
for manipulating FBMs, calculating statistics, and selecting genes.</p>
<p>Replaces all NaN and Inf values in a numeric vector with NA.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sparse_to_fbm</span><span class="op">(</span><span class="va">s_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">replace_with_na</span><span class="op">(</span><span class="va">v</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">normalize_matrix</span><span class="op">(</span></span>
<span>  <span class="va">x_mat</span>,</span>
<span>  normalization_method <span class="op">=</span> <span class="st">"divide_by_total_counts"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">get_anglemania_genes</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">get_anglemania_stats_df</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-s-mat">s_mat<a class="anchor" aria-label="anchor" href="#arg-s-mat"></a></dt>
<dd><p>A sparse matrix.</p></dd>


<dt id="arg-v">v<a class="anchor" aria-label="anchor" href="#arg-v"></a></dt>
<dd><p>A numeric vector.</p></dd>


<dt id="arg-x-mat">x_mat<a class="anchor" aria-label="anchor" href="#arg-x-mat"></a></dt>
<dd><p>A <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">bigstatsr::FBM</a></code> object containing the matrix to normalize
(typically genes x cells).</p></dd>


<dt id="arg-normalization-method">normalization_method<a class="anchor" aria-label="anchor" href="#arg-normalization-method"></a></dt>
<dd><p>A character string specifying the normalization
method to use. One of <code>"divide_by_total_counts"</code> (default) or
<code>"find_residuals"</code>.</p><ul><li><p><code>"divide_by_total_counts"</code> normalizes each cell by its total expression
count and applies log1p.</p></li>
<li><p><code>"find_residuals"</code> computes log1p-transformed residuals after regressing
out total expression.</p></li>
</ul></dd>


<dt id="arg-sce">sce<a class="anchor" aria-label="anchor" href="#arg-sce"></a></dt>
<dd><p>A SingleCellExperiment or SummarizedExperiment object</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>An <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code> object from the <span class="pkg">bigstatsr</span>
package.</p>
<p>A numeric vector with NaN and Inf values replaced with NA.</p>
<p>The input <code>FBM</code> object with normalized values written back in place.
This function modifies the input <code>x_mat</code> by reference.</p>
<p>A character vector of gene names that have been selected by the
anglemania algorithm</p>
<p>A data frame of gene pairs from which the anglemania genes
were selected</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>

<ul><li><p><code>sparse_to_fbm()</code>: Convert a sparse matrix into
a file-backed matrix
(<code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code>) with efficient memory usage.</p></li>
<li><p><code>replace_with_na()</code>: replace Nan and Inf values with NA</p></li>
<li><p><code>normalize_matrix()</code>: normalize matrix
Normalize a Filebacked Big Matrix (<code>FBM</code>) using either total-count
scaling or residuals from a linear model. Intended for use with
single-cell RNA-seq gene expression data.</p></li>
<li><p><code>get_anglemania_genes()</code>: Utility function to extract the genes that
have been selected by the anglemania algorithm.</p></li>
<li><p><code>get_anglemania_stats_df()</code>: Utility function to extract the stats of the
gene pairs from which the anglemania genes were selected.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">s_mat</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/rsparsematrix.html" class="external-link">rsparsematrix</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">10</span>, ncol <span class="op">=</span> <span class="fl">5</span>, density <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm_mat</span> <span class="op">&lt;-</span> <span class="fu">sparse_to_fbm</span><span class="op">(</span><span class="va">s_mat</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm_mat</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> A Filebacked Big Matrix of type 'double' with 10 rows and 5 columns.</span>
<span class="r-in"><span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="cn">Inf</span>, <span class="fl">9</span>, <span class="cn">NA</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu">replace_with_na</span><span class="op">(</span><span class="va">v</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">v</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1]  1  2  3  4  5  6  7 NA  9 NA</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigstatsr/" class="external-link">bigstatsr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">1000</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">100</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">as_FBM</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">normalize_matrix</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">fbm</span>,</span></span>
<span class="r-in"><span>  normalization_method <span class="op">=</span> <span class="st">"divide_by_total_counts"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [,1]     [,2]     [,3]     [,4]     [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 5.050104 4.789858 5.149273 4.605269 3.107610</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 5.167175 4.104990 4.682746 4.605269 4.680626</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 4.363345 4.104990 5.016571 3.703866 4.861401</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 4.917488 4.388541 4.682746 4.384622 4.459799</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 4.764556 5.192548 4.178070 5.071516 5.147147</span>
<span class="r-in"><span><span class="fu">normalize_matrix</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">fbm</span>,</span></span>
<span class="r-in"><span>  normalization_method <span class="op">=</span> <span class="st">"find_residuals"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            [,1]        [,2]        [,3]        [,4]          [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.11434410  0.05550312  0.10903721  0.01912041 -0.2880443664</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.09742030 -0.06396060  0.05569234  0.03692842  0.0434047229</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 0.01356898 -0.05693503  0.09785461 -0.14442728  0.0808357808</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 0.04243604 -0.01456288  0.05508198 -0.00548117 -0.0007333372</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 0.04158057  0.08936337 -0.10027542  0.06324354  0.0815543115</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing input...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing angles and transforming to z-scores...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s   </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/Rtmpf8hYYB/file2eeb697afe53" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |=========================                         | 50% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/Rtmpf8hYYB/file2eeb5caa0669" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing statistics...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Weighting matrix_list...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating mean...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating sds...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Pre-filtering features...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting filtered features...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selected 25 genes for integration.</span>
<span class="r-in"><span><span class="va">anglemania_genes</span> <span class="op">&lt;-</span> <span class="fu">get_anglemania_genes</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anglemania_genes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "gene43"  "gene207" "gene119" "gene130" "gene25"  "gene209"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">anglemania_genes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 25</span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Preparing input...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing angles and transforming to z-scores...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s   </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/Rtmpf8hYYB/file2eeb28fb7f08" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |=========================                         | 50% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/Rtmpf8hYYB/file2eeb297301f1" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing statistics...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Weighting matrix_list...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating mean...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating sds...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Pre-filtering features...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting filtered features...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selected 25 genes for integration.</span>
<span class="r-in"><span><span class="va">anglemania_stats_df</span> <span class="op">&lt;-</span> <span class="fu">get_anglemania_stats_df</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anglemania_stats_df</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     geneA   geneB sn_zscore mean_zscore</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  gene43 gene207 61.159210    3.345794</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 gene119 gene130  6.350243    3.014981</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  gene25 gene209  3.352897    2.983193</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  gene69 gene184  2.621413    2.858989</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  gene43  gene63  3.107353    2.735689</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  gene71 gene172  2.876975    2.691607</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">anglemania_stats_df</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 4</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Aaron Kollotzek, Vedran Franke, Artem Baranovskii, Altuna Akalin, SFB1588.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

