<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Utility Functions for the anglemania Package — anglemanise_utils • anglemania</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Utility Functions for the anglemania Package — anglemanise_utils"><meta property="og:description" content="A collection of utility functions used within the
anglemania package
for manipulating FBMs, calculating statistics, and selecting genes.
Replaces all NaN and Inf values in a numeric vector with NA."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">anglemania</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/anglemania_tutorial_Seurat.html">Running anglemania with Seurat</a>
    </li>
    <li>
      <a href="../articles/anglemania_tutorial.html">Running anglemania with SingleCellExperiment</a>
    </li>
    <li>
      <a href="../articles/Functions_Showcasing.html">Functions_Showcasing</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Utility Functions for the anglemania Package</h1>
    <small class="dont-index">Source: <a href="https://github.com/BIMSBbioinfo/anglemania/blob/master/R/anglemanise_utils.R" class="external-link"><code>R/anglemanise_utils.R</code></a></small>
    <div class="hidden name"><code>anglemanise_utils.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A collection of utility functions used within the
<span class="pkg">anglemania</span> package
for manipulating FBMs, calculating statistics, and selecting genes.</p>
<p>Replaces all NaN and Inf values in a numeric vector with NA.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sparse_to_fbm</span><span class="op">(</span><span class="va">s_mat</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">get_dstat</span><span class="op">(</span><span class="va">corr_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">big_mat_list_mean</span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">get_list_stats</span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">extract_rows_for_unique_genes</span><span class="op">(</span><span class="va">dt</span>, <span class="va">max_n_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">prefilter_angl</span><span class="op">(</span><span class="va">angl</span>, zscore_mean_threshold <span class="op">=</span> <span class="fl">1</span>, zscore_sn_threshold <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">select_genes</span><span class="op">(</span></span>
<span>  <span class="va">angl</span>,</span>
<span>  zscore_mean_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  zscore_sn_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  max_n_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  direction <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  adjust_thresholds <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">replace_with_na</span><span class="op">(</span><span class="va">v</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">variable_genes_overlap</span><span class="op">(</span></span>
<span>  <span class="va">seu</span>,</span>
<span>  <span class="va">angl</span>,</span>
<span>  zscore_mean_thresholds <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zscore_sn_thresholds <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  adjust_thresholds <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  length_out <span class="op">=</span> <span class="fl">6</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">normalize_matrix</span><span class="op">(</span><span class="va">x_mat</span>, normalization_method <span class="op">=</span> <span class="st">"divide_by_total_counts"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-s-mat">s_mat<a class="anchor" aria-label="anchor" href="#arg-s-mat"></a></dt>
<dd><p>A sparse matrix.</p></dd>


<dt id="arg-corr-matrix">corr_matrix<a class="anchor" aria-label="anchor" href="#arg-corr-matrix"></a></dt>
<dd><p>An <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code> object.</p></dd>


<dt id="arg-angl">angl<a class="anchor" aria-label="anchor" href="#arg-angl"></a></dt>
<dd><p>An anglemania analysis object. This is used by <code>select_genes</code>
and <code>get_anglemania_genes</code> to select genes based on the provided thresholds.</p></dd>


<dt id="arg-dt">dt<a class="anchor" aria-label="anchor" href="#arg-dt"></a></dt>
<dd><p>A data frame containing gene pairs, with columns <code>geneA</code>
and <code>geneB</code>.</p></dd>


<dt id="arg-max-n-genes">max_n_genes<a class="anchor" aria-label="anchor" href="#arg-max-n-genes"></a></dt>
<dd><p>Integer specifying the maximum number of genes to select.
If <code>NULL</code>, all genes that pass the thresholds are used. Default is
<code>NULL</code>.</p></dd>


<dt id="arg-zscore-mean-threshold">zscore_mean_threshold<a class="anchor" aria-label="anchor" href="#arg-zscore-mean-threshold"></a></dt>
<dd><p>Numeric value specifying the threshold for the
absolute mean z-score. Default is 2.</p></dd>


<dt id="arg-zscore-sn-threshold">zscore_sn_threshold<a class="anchor" aria-label="anchor" href="#arg-zscore-sn-threshold"></a></dt>
<dd><p>Numeric value specifying the threshold for the
SNR z-score. Default is 2.</p></dd>


<dt id="arg-direction">direction<a class="anchor" aria-label="anchor" href="#arg-direction"></a></dt>
<dd><p>whether to select genes with positive, negative or both
mean z-scores. Default is "both"</p></dd>


<dt id="arg-adjust-thresholds">adjust_thresholds<a class="anchor" aria-label="anchor" href="#arg-adjust-thresholds"></a></dt>
<dd><p>Logical. Whether to use adaptive thresholding in
<code>select_genes</code>. Default is FALSE.</p></dd>


<dt id="arg-v">v<a class="anchor" aria-label="anchor" href="#arg-v"></a></dt>
<dd><p>A numeric vector.</p></dd>


<dt id="arg-seu">seu<a class="anchor" aria-label="anchor" href="#arg-seu"></a></dt>
<dd><p>A Seurat object. If variable features have not been identified,
they will be computed using <code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></code>.</p></dd>


<dt id="arg-zscore-mean-thresholds">zscore_mean_thresholds<a class="anchor" aria-label="anchor" href="#arg-zscore-mean-thresholds"></a></dt>
<dd><p>A numeric vector of zscore mean thresholds used
for gene selection from <code>ang_obj</code>. Default is <code>c(0.5, 1:15)</code>.</p></dd>


<dt id="arg-zscore-sn-thresholds">zscore_sn_thresholds<a class="anchor" aria-label="anchor" href="#arg-zscore-sn-thresholds"></a></dt>
<dd><p>A numeric vector of zscore signal to noise ratio
thresholds used for gene selection from <code>ang_obj</code>. Default is
<code>c(0.5, 1:15)</code>.</p></dd>


<dt id="arg-layer">layer<a class="anchor" aria-label="anchor" href="#arg-layer"></a></dt>
<dd><p>Character. The assay layer in the Seurat object to use when
calculating variable features (e.g., "counts", "data", or "logcounts").
Default is <code>"data"</code>.</p></dd>


<dt id="arg-x-mat">x_mat<a class="anchor" aria-label="anchor" href="#arg-x-mat"></a></dt>
<dd><p>A <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">bigstatsr::FBM</a></code> object containing the matrix to normalize
(typically genes x cells).</p></dd>


<dt id="arg-normalization-method">normalization_method<a class="anchor" aria-label="anchor" href="#arg-normalization-method"></a></dt>
<dd><p>A character string specifying the normalization
method to use. One of <code>"divide_by_total_counts"</code> (default) or
<code>"find_residuals"</code>.</p><ul><li><p><code>"divide_by_total_counts"</code> normalizes each cell by its total expression
count and applies log1p.</p></li>
<li><p><code>"find_residuals"</code> computes log1p-transformed residuals after regressing
out total expression.</p></li>
</ul></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>An <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code> object from the <span class="pkg">bigstatsr</span>
package.</p>
<p>A list with statistical measures including <code>mean</code>, <code>sd</code>,
<code>var</code>, <code>min</code>, and <code>max</code>.</p>
<p>A new <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code> object containing the mean values.</p>
<p>A list containing three matrices: <code>mean_zscore</code>,
<code>sds_zscore</code>, and <code>sn_zscore</code>.</p>
<p>A vector of unique gene identifiers.</p>
<p>The input <code>anglemania_object</code> with the
<code>integration_genes</code> slot updated to include the selected genes and
their statistical information.</p>
<p>The input <code>anglemaniaObject</code> with the
<code>integration_genes</code> slot updated to include the selected genes and
their statistical information.</p>
<p>A numeric vector with NaN and Inf values replaced with NA.</p>
<p>A data frame with the following columns:</p><dl><dt>threshold</dt>
<dd><p>The threshold value used for gene selection.</p></dd>

<dt>intersecting_genes</dt>
<dd><p>The number of genes that are both variable in
<code>seu</code> and selected by <code>ang_obj</code>.</p></dd>

<dt>perc_var_genes</dt>
<dd><p>The proportion of variable genes in <code>seu</code> that
are also selected by <code>ang_obj</code>.</p></dd>

<dt>number_angl_genes</dt>
<dd><p>The total number of genes selected by
<code>ang_obj</code> at the given threshold.</p></dd>


</dl><p>The input <code>FBM</code> object with normalized values written back in place.
This function modifies the input <code>x_mat</code> by reference.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function combines the <code>geneA</code> and <code>geneB</code> columns, extracts
unique gene names, and returns the first <code>max_n_genes</code> genes. If
<code>max_n_genes</code> exceeds the number of unique genes available, all unique
genes are returned.</p>
<p>The function performs the following steps:</p><ol><li><p>Checks if the input object is of class <code><a href="anglemania_object-class.html">anglemania_object-class</a></code>.</p></li>
<li><p>Identifies gene pairs where both the mean z-score and SNR z-score
exceed the specified thresholds.</p></li>
</ol><p>The function performs the following steps:</p><ol><li><p>Checks if the input object is of class <code><a href="anglemania_object-class.html">anglemania_object-class</a></code>.</p></li>
<li><p>If <code>max_n_genes</code> is not specified, it uses all genes that pass
the thresholds.</p></li>
<li><p>Identifies gene pairs where both the mean z-score and SNR z-score
exceed the specified thresholds.</p></li>
<li><p>If no gene pairs meet the criteria, it adjusts the thresholds to the
99th percentile values of the corresponding statistics and re-selects.</p></li>
<li><p>Extracts unique genes from the selected gene pairs using
<code>extract_rows_for_unique_genes</code>.</p></li>
<li><p>Updates the <code>integration_genes</code> slot of the
<code>anglemania_object</code>
with the selected genes and their statistics.</p></li>
</ol></div>
    <div id="functions">
    <h2>Functions</h2>

<ul><li><p><code>sparse_to_fbm()</code>: Convert a sparse matrix into a file-backed matrix
(<code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code>) with efficient memory usage.</p></li>
<li><p><code>get_dstat()</code>: Compute mean, standard deviation, variance,
min, and max of a correlation matrix stored as an
<code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code>.</p></li>
<li><p><code>big_mat_list_mean()</code>: Calculates the element-wise mean from a list
of <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code>s stored in an <code>anglemania_object</code>.</p></li>
<li><p><code>get_list_stats()</code>: Calculate mean, standard deviation, and SNR
across a list of FBMs stored in an <code>anglemania_object</code>.</p></li>
<li><p><code>extract_rows_for_unique_genes()</code>: Extract unique gene identifiers from gene pairs,
returning up to a specified maximum number.</p></li>
<li><p><code>prefilter_angl()</code>: Preselect genes from an
<code>anglemania_object</code>
based on z-score thresholds, to simplify downstream filtering.</p></li>
<li><p><code>select_genes()</code>: Select genes from an <code>anglemania_object</code>
based on z-score thresholds for mean and signal-to-noise ratio (SNR).
Selects genes from an <code><a href="anglemania_object-class.html">anglemania_object-class</a></code> based on specified
thresholds for the absolute mean z-score and signal-to-noise ratio
(SNR) z-score. It updates the <code>integration_genes</code> slot of the
<code>anglemania_object</code> with the selected genes and associated
information.</p></li>
<li><p><code>replace_with_na()</code>: replace Nan and Inf values with NA</p></li>
<li><p><code>variable_genes_overlap()</code>: Overlap of Variable Genes Across Thresholds.
This function computes the overlap between variable genes identified in a
Seurat object and the gene sets selected from an "anglemania" analysis
object (<code>ang_obj</code>) at multiple threshold levels. For each threshold, genes
are selected from <code>ang_obj</code> by applying <code>select_genes</code>, and the intersection
of these genes with the variable features of the Seurat object is calculated.
The result is a summary data frame that includes the number and percentage of
variable genes overlapping with the anglemania-selected genes, as well as the
number of genes selected at each threshold.</p></li>
<li><p><code>normalize_matrix()</code>: normalize matrix
Normalize a Filebacked Big Matrix (<code>FBM</code>) using either total-count
scaling or residuals from a linear model. Intended for use with
single-cell RNA-seq gene expression data.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="https://privefl.github.io/bigstatsr/reference/big_apply.html" class="external-link">big_apply</a></code>, <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code></p>
<p><code><a href="https://privefl.github.io/bigstatsr/reference/big_apply.html" class="external-link">big_apply</a></code>, <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code></p>
<p><code>select_genes</code></p>
<p><code>extract_rows_for_unique_genes</code>,
<code><a href="anglemania_object-methods.html">intersect_genes</a></code>, <code><a href="anglemania_object-methods.html">list_stats</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">s_mat</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/rsparsematrix.html" class="external-link">rsparsematrix</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">10</span>, ncol <span class="op">=</span> <span class="fl">5</span>, density <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm_mat</span> <span class="op">&lt;-</span> <span class="fu">sparse_to_fbm</span><span class="op">(</span><span class="va">s_mat</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm_mat</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> A Filebacked Big Matrix of type 'double' with 10 rows and 5 columns.</span>
<span class="r-in"><span><span class="va">s_mat</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/rsparsematrix.html" class="external-link">rsparsematrix</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">10</span>, ncol <span class="op">=</span> <span class="fl">5</span>, density <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm_mat</span> <span class="op">&lt;-</span> <span class="fu">sparse_to_fbm</span><span class="op">(</span><span class="va">s_mat</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">get_dstat</span><span class="op">(</span><span class="va">fbm_mat</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ mean: num [1:5] -0.033 -0.211 0.209 0 0.083</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ var : num [1:5] 0.1126 0.2017 0.2051 0.0827 0.5553</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ min : num [1:5] -0.56 -1.4 -0.27 -0.61 -1.2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ max : num [1:5] 0.7 0 1.2 0.61 1.9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ sd  : num [1:5] 0.336 0.449 0.453 0.288 0.745</span>
<span class="r-in"><span><span class="va">result</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $mean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] -0.033 -0.211  0.209  0.000  0.083</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $var</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.11262333 0.20165444 0.20507667 0.08268889 0.55533444</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $min</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] -0.56 -1.40 -0.27 -0.61 -1.20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $max</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.70 0.00 1.20 0.61 1.90</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $sd</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.3355940 0.4490595 0.4528539 0.2875568 0.7452077</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="co"># Create FBMs</span></span></span>
<span class="r-in"><span><span class="va">mat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, nrow <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fbm1</span> <span class="op">&lt;-</span> <span class="fu">bigstatsr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mat1</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">mat1</span><span class="op">)</span>, init <span class="op">=</span> <span class="va">mat1</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm2</span> <span class="op">&lt;-</span> <span class="fu">bigstatsr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mat2</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">mat2</span><span class="op">)</span>, init <span class="op">=</span> <span class="va">mat2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create weights</span></span></span>
<span class="r-in"><span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>batch1 <span class="op">=</span> <span class="fl">0.5</span>, batch2 <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create the list of FBMs</span></span></span>
<span class="r-in"><span><span class="va">fbm_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>batch1 <span class="op">=</span> <span class="va">fbm1</span>, batch2 <span class="op">=</span> <span class="va">fbm2</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Construct the anglemania_object</span></span></span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"anglemania_object"</span>,</span></span>
<span class="r-in"><span>  weights <span class="op">=</span> <span class="va">weights</span>,</span></span>
<span class="r-in"><span>  matrix_list <span class="op">=</span> <span class="va">fbm_list</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">big_mat_list_mean</span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> A Filebacked Big Matrix of type 'double' with 3 rows and 3 columns.</span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_anglemania_object.html">create_anglemania_object</a></span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing angles and transforming to z-scores...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s   </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/RtmpCUFAZZ/file20ab1b42ed1c" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |=========================                         | 50% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/RtmpCUFAZZ/file20ab5b40a14c" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing statistics...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Weighting matrix_list...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating mean...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating sds...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering features...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selecting features...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 25 genes for integration."</span>
<span class="r-in"><span><span class="fu"><a href="anglemania_object-methods.html">list_stats</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">get_list_stats</span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Weighting matrix_list...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating mean...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating sds...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="anglemania_object-methods.html">list_stats</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ mean_zscore:Reference class 'FBM' [package "bigstatsr"] with 15 fields</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr      :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr_rw   :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ nrow        : int 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ ncol        : int 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type        : Named int 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..- attr(*, "names")= chr "double"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ backingfile : chr "/tmp/RtmpCUFAZZ/file20ab6bfb4a5a.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_read_only: logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address     :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address_rw  :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ bk          : chr "/tmp/RtmpCUFAZZ/file20ab6bfb4a5a.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ rds         : chr "/tmp/RtmpCUFAZZ/file20ab6bfb4a5a.rds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_saved    : logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_chr    : chr "double"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_size   : int 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ file_size   : num 720000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..and 22 methods, of which 8 are  possibly relevant:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  add_columns, bm, bm.desc, check_dimensions, check_write_permissions,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  initialize, save, show#envRefClass</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ sds_zscore :Reference class 'FBM' [package "bigstatsr"] with 15 fields</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr      :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr_rw   :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ nrow        : int 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ ncol        : int 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type        : Named int 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..- attr(*, "names")= chr "double"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ backingfile : chr "/tmp/RtmpCUFAZZ/file20ab1e3da11.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_read_only: logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address     :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address_rw  :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ bk          : chr "/tmp/RtmpCUFAZZ/file20ab1e3da11.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ rds         : chr "/tmp/RtmpCUFAZZ/file20ab1e3da11.rds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_saved    : logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_chr    : chr "double"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_size   : int 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ file_size   : num 720000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..and 22 methods, of which 8 are  possibly relevant:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  add_columns, bm, bm.desc, check_dimensions, check_write_permissions,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  initialize, save, show#envRefClass</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ sn_zscore  :Reference class 'FBM' [package "bigstatsr"] with 15 fields</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr      :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ extptr_rw   :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ nrow        : int 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ ncol        : int 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type        : Named int 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..- attr(*, "names")= chr "double"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ backingfile : chr "/tmp/RtmpCUFAZZ/file20ab717897d4.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_read_only: logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address     :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ address_rw  :&lt;externalptr&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ bk          : chr "/tmp/RtmpCUFAZZ/file20ab717897d4.bk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ rds         : chr "/tmp/RtmpCUFAZZ/file20ab717897d4.rds"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ is_saved    : logi FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_chr    : chr "double"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ type_size   : int 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ file_size   : num 720000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..and 22 methods, of which 8 are  possibly relevant:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  add_columns, bm, bm.desc, check_dimensions, check_write_permissions,</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..  initialize, save, show#envRefClass</span>
<span class="r-in"><span><span class="va">gene_pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  geneA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>, <span class="st">"Gene3"</span>, <span class="st">"Gene4"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  geneB <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene3"</span>, <span class="st">"Gene4"</span>, <span class="st">"Gene5"</span>, <span class="st">"Gene6"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">unique_genes</span> <span class="op">&lt;-</span> <span class="fu">extract_rows_for_unique_genes</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gene_pairs</span>,</span></span>
<span class="r-in"><span>  max_n_genes <span class="op">=</span> <span class="fl">3</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">unique_genes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Gene1" "Gene3" "Gene2"</span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_anglemania_object.html">create_anglemania_object</a></span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing angles and transforming to z-scores...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s   </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/RtmpCUFAZZ/file20ab6fd65dbf" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |=========================                         | 50% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/RtmpCUFAZZ/file20abf2fda0" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing statistics...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Weighting matrix_list...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating mean...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating sds...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering features...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selecting features...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 25 genes for integration."</span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu">select_genes</span><span class="op">(</span><span class="va">angl</span>,</span></span>
<span class="r-in"><span>                      zscore_mean_threshold <span class="op">=</span> <span class="fl">2.5</span>,</span></span>
<span class="r-in"><span>                     zscore_sn_threshold <span class="op">=</span> <span class="fl">2.5</span>,</span></span>
<span class="r-in"><span>                     max_n_genes <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 25 genes for integration."</span>
<span class="r-in"><span><span class="va">anglemania_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="anglemania_object-methods.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># View the selected genes and use for integration</span></span></span>
<span class="r-in"><span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="cn">Inf</span>, <span class="fl">9</span>, <span class="cn">NA</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu">replace_with_na</span><span class="op">(</span><span class="va">v</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">v</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1]  1  2  3  4  5  6  7 NA  9 NA</span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Data is of class matrix. Coercing to dgCMatrix.</span>
<span class="r-in"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding variable features for layer counts</span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_anglemania_object.html">create_anglemania_object</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">sce</span>,</span></span>
<span class="r-in"><span>  batch_key <span class="op">=</span> <span class="st">"batch"</span>,</span></span>
<span class="r-in"><span>  min_cells_per_gene <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing angles and transforming to z-scores...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s   </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/RtmpCUFAZZ/file20ab45ae24bf" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |=========================                         | 50% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating directory "/tmp/RtmpCUFAZZ/file20ab4120f4a6" which didn't exist..</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing statistics...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Weighting matrix_list...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating mean...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating sds...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering features...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Selecting features...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 25 genes for integration."</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">variable_genes_overlap</span><span class="op">(</span><span class="va">seu</span>, <span class="va">angl</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 300 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 300 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 300 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 119 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 11 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 2 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 4 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 4 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Selected 1 genes for integration."</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   zscore_mean_thresholds zscore_sn_thresholds intersecting_genes perc_var_genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1               0.500000                  0.5                300    1.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2               1.101441                  0.5                300    1.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3               1.702882                  0.5                300    1.000000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4               2.304323                  0.5                119    0.396666667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5               2.905764                  0.5                 11    0.036666667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6               3.507205                  0.5                  2    0.006666667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   number_angl_genes perc_ang_genes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1               300              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2               300              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3               300              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4               119              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                11              1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                 2              1</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://privefl.github.io/bigstatsr/" class="external-link">bigstatsr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">1000</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">100</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fbm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">as_FBM</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">normalize_matrix</span><span class="op">(</span><span class="va">fbm</span>, normalization_method <span class="op">=</span> <span class="st">"divide_by_total_counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [,1]     [,2]     [,3]     [,4]     [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 5.050104 4.789858 5.149273 4.605269 3.107610</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 5.167175 4.104990 4.682746 4.605269 4.680626</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 4.363345 4.104990 5.016571 3.703866 4.861401</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 4.917488 4.388541 4.682746 4.384622 4.459799</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 4.764556 5.192548 4.178070 5.071516 5.147147</span>
<span class="r-in"><span><span class="fu">normalize_matrix</span><span class="op">(</span><span class="va">fbm</span>, normalization_method <span class="op">=</span> <span class="st">"find_residuals"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            [,1]        [,2]        [,3]        [,4]          [,5]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.11434410  0.05550312  0.10903721  0.01912041 -0.2880443664</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.09742030 -0.06396060  0.05569234  0.03692842  0.0434047229</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 0.01356898 -0.05693503  0.09785461 -0.14442728  0.0808357808</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 0.04243604 -0.01456288  0.05508198 -0.00548117 -0.0007333372</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 0.04158057  0.08936337 -0.10027542  0.06324354  0.0815543115</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Aaron Kollotzek, Vedran Franke, Artem Baranovskii, Altuna Akalin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

