<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create an anglemania_object from a Seurat or SingleCellExperiment Object — create_anglemania_object • anglemania</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create an anglemania_object from a Seurat or SingleCellExperiment Object — create_anglemania_object"><meta property="og:description" content="Constructs an anglemania_object-class from a given
Seurat or
SingleCellExperiment object.
This includes extracting and processing count matrices, filtering genes
based on expression in a minimum number of cells, and storing results
along with dataset and batch information. It also calculates weights for
each dataset or batch based on the number of samples."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">anglemania</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/anglemania_tutorial_Seurat.html">Running anglemania with Seurat</a>
    </li>
    <li>
      <a href="../articles/anglemania_tutorial.html">Running anglemania with SingleCellExperiment</a>
    </li>
    <li>
      <a href="../articles/Functions_Showcasing.html">Functions_Showcasing</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create an anglemania_object from a Seurat or SingleCellExperiment Object</h1>
    <small class="dont-index">Source: <a href="https://github.com/BIMSBbioinfo/anglemania/blob/master/R/objects.R" class="external-link"><code>R/objects.R</code></a></small>
    <div class="hidden name"><code>create_anglemania_object.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Constructs an <code><a href="anglemania_object-class.html">anglemania_object-class</a></code> from a given
<code><a href="https://satijalab.org/seurat/reference/Seurat-package.html" class="external-link">Seurat</a></code> or
<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object.
This includes extracting and processing count matrices, filtering genes
based on expression in a minimum number of cells, and storing results
along with dataset and batch information. It also calculates weights for
each dataset or batch based on the number of samples.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_anglemania_object</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'Seurat'</span></span>
<span><span class="fu">create_anglemania_object</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">batch_key</span>,</span>
<span>  min_cells_per_gene <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  dataset_key <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SingleCellExperiment'</span></span>
<span><span class="fu">create_anglemania_object</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">batch_key</span>,</span>
<span>  min_cells_per_gene <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  dataset_key <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'list'</span></span>
<span><span class="fu">create_anglemania_object</span><span class="op">(</span><span class="va">object</span>, min_cells_per_gene <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A <code><a href="https://satijalab.org/seurat/reference/Seurat-package.html" class="external-link">Seurat</a></code> or
<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code> object
containing single-cell RNA-seq data.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments</p></dd>


<dt id="arg-batch-key">batch_key<a class="anchor" aria-label="anchor" href="#arg-batch-key"></a></dt>
<dd><p>A character string indicating the column name(s) in the
object metadata that identify the batch to which each cell belongs.</p></dd>


<dt id="arg-min-cells-per-gene">min_cells_per_gene<a class="anchor" aria-label="anchor" href="#arg-min-cells-per-gene"></a></dt>
<dd><p>A numeric value indicating the minimum number of
cells in which a gene must be expressed to be included in the analysis.
Default is <code>1</code>.</p></dd>


<dt id="arg-dataset-key">dataset_key<a class="anchor" aria-label="anchor" href="#arg-dataset-key"></a></dt>
<dd><p>A character string indicating the column name in the
object metadata that identifies the dataset to which each cell
belongs. If <code>NA</code>, all cells are assumed to belong to the same
dataset.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>An <code><a href="anglemania_object-class.html">anglemania_object-class</a></code> containing:</p><dl><dt><code>matrix_list</code></dt>
<dd><p>A list of filtered count matrices for each unique
batch.</p></dd>

<dt><code>dataset_key</code></dt>
<dd><p>The dataset key used for splitting the
object.</p></dd>

<dt><code>batch_key</code></dt>
<dd><p>The batch key used for splitting the
object.</p></dd>

<dt><code>data_info</code></dt>
<dd><p>A data frame summarizing the number of samples per
dataset and their weights.</p></dd>

<dt><code>weights</code></dt>
<dd><p>A numeric vector of weights for each dataset or batch
based on the number of samples.</p></dd>

<dt><code>intersect_genes</code></dt>
<dd><p>A character vector of genes expressed in at
least the specified number of cells across all batches.</p></dd>

<dt><code>min_cells_per_gene</code></dt>
<dd><p>The minimum number of cells per gene
threshold used for filtering.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>The function performs the following steps:</p><ol><li><p>Adds a unique batch key to the metadata using
<code><a href="anglemania_object-methods.html">add_unique_batch_key</a></code>.</p></li>
<li><p>Extracts count matrices for each batch.</p></li>
<li><p>Filters genes based on the <code>min_cells_per_gene</code> threshold.</p></li>
<li><p>Identifies intersected genes present across all batches.</p></li>
<li><p>Converts count matrices to <code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code> objects.</p></li>
<li><p>Computes weights for each batch or dataset.</p></li>
</ol></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="anglemania_object-class.html">anglemania_object-class</a></code>,
<code><a href="anglemania_object-methods.html">add_unique_batch_key</a></code>,
<code><a href="anglemania.html">anglemania</a></code>,
<code><a href="https://privefl.github.io/bigstatsr/reference/FBM-class.html" class="external-link">FBM</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu">create_anglemania_object</span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> anglemania_object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dataset key: NA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batch key: batch </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of datasets: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of batches: 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batches (showing first 5):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> batch1, batch2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of intersected genes: 300 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Intersected genes (showing first 10):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1, gene2, gene3, gene4, gene5, gene6, gene7, gene8, gene9, gene10 , ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Min cells per gene: 1 </span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">sce</span>, data <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/RenameAssays.html" class="external-link">RenameAssays</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"originalexp"</span>, <span class="st">"RNA"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Renaming default assay from originalexp to RNA</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Key ‘originalexp_’ taken, using ‘rna_’ instead</span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu">create_anglemania_object</span><span class="op">(</span><span class="va">se</span>, batch_key <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> anglemania_object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dataset key: NA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batch key: batch </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of datasets: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of batches: 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batches (showing first 5):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> batch1, batch2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of intersected genes: 300 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Intersected genes (showing first 10):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1, gene2, gene3, gene4, gene5, gene6, gene7, gene8, gene9, gene10 , ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Min cells per gene: 1 </span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu">create_anglemania_object</span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="st">"batch"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dataset_key specified.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> anglemania_object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dataset key: NA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batch key: batch </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of datasets: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of batches: 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batches (showing first 5):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> batch1, batch2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of intersected genes: 300 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Intersected genes (showing first 10):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1, gene2, gene3, gene4, gene5, gene6, gene7, gene8, gene9, gene10 , ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Min cells per gene: 1 </span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sce1 <span class="op">=</span> <span class="va">sce</span>, sce2 <span class="op">=</span> <span class="va">sce</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu">create_anglemania_object</span><span class="op">(</span><span class="va">sce_list</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> anglemania_object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dataset key: NA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batch key: batch </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of datasets: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of batches: 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batches (showing first 5):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sce1, sce2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of intersected genes: 300 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Intersected genes (showing first 10):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1, gene2, gene3, gene4, gene5, gene6, gene7, gene8, gene9, gene10 , ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Min cells per gene: 1 </span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">sce</span>, data <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/RenameAssays.html" class="external-link">RenameAssays</a></span><span class="op">(</span><span class="va">se</span>, <span class="st">"originalexp"</span>, <span class="st">"RNA"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Renaming default assay from originalexp to RNA</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Key ‘originalexp_’ taken, using ‘rna_’ instead</span>
<span class="r-in"><span><span class="va">se_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>se1 <span class="op">=</span> <span class="va">se</span>, se2 <span class="op">=</span> <span class="va">se</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu">create_anglemania_object</span><span class="op">(</span><span class="va">se_list</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extracting count matrices...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering each batch to at least 1 cells per gene...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the intersection of filtered genes from all batches...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of genes in intersected set: 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % elapsed=00s     |==================================================| 100% elapsed=00s, remaining~00s</span>
<span class="r-in"><span><span class="va">angl</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> anglemania_object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> --------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dataset key: NA </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batch key: batch </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of datasets: 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of batches: 2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Batches (showing first 5):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> se1, se2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of intersected genes: 300 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Intersected genes (showing first 10):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1, gene2, gene3, gene4, gene5, gene6, gene7, gene8, gene9, gene10 , ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Min cells per gene: 1 </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Aaron Kollotzek, Vedran Franke, Artem Baranovskii, Altuna Akalin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

