<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Running anglemania with SingleCellExperiment • anglemania</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Running anglemania with SingleCellExperiment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">anglemania</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/anglemania_tutorial_Seurat.html">Running anglemania with Seurat</a>
    </li>
    <li>
      <a href="../articles/anglemania_tutorial.html">Running anglemania with SingleCellExperiment</a>
    </li>
    <li>
      <a href="../articles/Functions_Showcasing.html">Functions_Showcasing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Running anglemania with
SingleCellExperiment</h1>
            
            <h4 data-toc-skip class="date">2025-04-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BIMSBbioinfo/anglemania/blob/master/vignettes/anglemania_tutorial.Rmd" class="external-link"><code>vignettes/anglemania_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>anglemania_tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>anglemania is a feature selection package that extracts genes from
multi-batch scRNA-seq experiments for downstream dataset integration.
The goal is to select genes that carry high biological information and
low technical noise between the batches. Those genes are extracted from
gene pairs that have an invariant and extremely narrow or wide angle
between their expression vectors. Conventionally, highly-variable genes
(HVGs) or sometimes all genes are used for integration tasks (<a href="https://www.nature.com/articles/s41592-021-01336-8" class="external-link uri">https://www.nature.com/articles/s41592-021-01336-8</a>).
While HVGs are a great and easy way to reduce the noise and
dimensionality of the data, they are not optimal for integration tasks.
HVGs are sensitive to batch effects because the variance is a function
of both the technical and biological variance. anglemania improves
conventional usage of HVGs for integration tasks, especially when the
transcriptional difference between cell types or cell states is subtle
(showcased here with de.facLoc set to 0.1 (= mild differences between
“Groups”)). The package can be used on top of SingleCellExperiment or
Seurat objects.</p>
<p>Under the hood, anglemania works with file-backed big matrices (FBMs)
from the bigstatsr package (<a href="https://github.com/privefl/bigstatsr/" class="external-link uri">https://github.com/privefl/bigstatsr/</a>) for fast and
memory efficient computation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinformatics.mdc-berlin.de/anglemania" class="external-link">anglemania</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/splatter/" class="external-link">splatter</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by</span></span>
<span><span class="co">## 'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="create-simulated-data-with-mild-batch-effects-and-mild-differences-between-the-groupscell-types">Create simulated data with mild batch effects and mild differences
between the groups/cell types<a class="anchor" aria-label="anchor" href="#create-simulated-data-with-mild-batch-effects-and-mild-differences-between-the-groupscell-types"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch.facLoc</span> <span class="op">&lt;-</span> <span class="fl">0.3</span></span>
<span><span class="va">de.facLoc</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">nBatches</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="va">nGroups</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span>
<span><span class="va">groupCells</span> <span class="op">&lt;-</span> <span class="fl">300</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://oshlack.github.io/splatter/reference/splatSimulate.html" class="external-link">splatSimulate</a></span><span class="op">(</span></span>
<span>  batchCells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">groupCells</span> <span class="op">*</span> <span class="va">nGroups</span>, <span class="va">nBatches</span><span class="op">)</span>,</span>
<span>  batch.facLoc <span class="op">=</span> <span class="va">batch.facLoc</span>,</span>
<span>  group.prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="va">nGroups</span>, <span class="va">nGroups</span><span class="op">)</span>,</span>
<span>  nGenes <span class="op">=</span> <span class="va">nGenes</span>,</span>
<span>  batch.facScale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"groups"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  out.prob <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  de.prob <span class="op">=</span> <span class="fl">0.1</span>, <span class="co"># mild</span></span>
<span>  de.facLoc <span class="op">=</span> <span class="va">de.facLoc</span>,</span>
<span>  de.facScale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  bcv.common <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">42</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 5000 3600 </span></span>
<span><span class="co">## metadata(1): Params</span></span>
<span><span class="co">## assays(6): BatchCellMeans BaseCellMeans ... TrueCounts counts</span></span>
<span><span class="co">## rownames(5000): Gene1 Gene2 ... Gene4999 Gene5000</span></span>
<span><span class="co">## rowData names(11): Gene BaseGeneMean ... DEFacGroup2 DEFacGroup3</span></span>
<span><span class="co">## colnames(3600): Cell1 Cell2 ... Cell3599 Cell3600</span></span>
<span><span class="co">## colData names(4): Cell Batch Group ExpLibSize</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 6</span></span>
<span><span class="co">## names(6): BatchCellMeans BaseCellMeans BCV CellMeans TrueCounts counts</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="unintegrated-data">unintegrated data<a class="anchor" aria-label="anchor" href="#unintegrated-data"></a>
</h2>
<ul>
<li>clusters are driven by batch effects</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="va">sce</span></span>
<span><span class="co"># Normalization.</span></span>
<span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce_unintegrated</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Feature selection.</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce_unintegrated</span><span class="op">)</span></span>
<span><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, prop <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>, ncomponents <span class="op">=</span> <span class="fl">50</span>, subset_row <span class="op">=</span> <span class="va">hvg</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Clustering.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce_unintegrated</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>,</span>
<span>  use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>  BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualization.</span></span>
<span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-1-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="run-anglemania">run anglemania<a class="anchor" aria-label="anchor" href="#run-anglemania"></a>
</h2>
<div class="section level3">
<h3 id="create-anglemania-object">create anglemania object<a class="anchor" aria-label="anchor" href="#create-anglemania-object"></a>
</h3>
<ul>
<li>We create an anglemania object to have a standardized input to the
anglemania function
<ul>
<li>It can be created from:
<ul>
<li>An SingleCellExperiment (SCE) object</li>
<li>Seurat object</li>
<li>a list of SCE and/or Seurat objects</li>
</ul>
</li>
</ul>
</li>
<li>pre-requisites are that you:
<ul>
<li>have a column in the metadata (colData for SCE) that specifies the
batch/sample a cell belongs to</li>
<li>or if you pass a list of SCE/Seurat objects, just provide this as
the input and don’t specify a batch key</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">##              Cell       Batch    Group ExpLibSize</span></span>
<span><span class="co">##       &lt;character&gt; &lt;character&gt; &lt;factor&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## Cell1       Cell1      Batch1   Group1    46898.1</span></span>
<span><span class="co">## Cell2       Cell2      Batch1   Group1    54688.2</span></span>
<span><span class="co">## Cell3       Cell3      Batch1   Group2    52027.9</span></span>
<span><span class="co">## Cell4       Cell4      Batch1   Group1    52319.5</span></span>
<span><span class="co">## Cell5       Cell5      Batch1   Group3    37774.7</span></span>
<span><span class="co">## Cell6       Cell6      Batch1   Group3    58112.3</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch_key</span> <span class="op">&lt;-</span> <span class="st">"Batch"</span></span>
<span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_anglemania_object.html">create_anglemania_object</a></span><span class="op">(</span></span>
<span>  <span class="va">sce</span>,</span>
<span>  batch_key <span class="op">=</span> <span class="va">batch_key</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## No dataset_key specified.</span></span>
<span><span class="co">## Assuming that all samples belong to the same dataset and are separated by batch_key: Batch</span></span></code></pre>
<pre><code><span><span class="co">## Using the intersection of filtered genes from all batches...</span></span></code></pre>
<pre><code><span><span class="co">## Number of genes in intersected set: 4964</span></span></code></pre>
<pre><code><span><span class="co">## Extracting count matrices...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering each batch to at least 1 cells per gene...</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span></span></code></pre></div>
<pre><code><span><span class="co">## anglemania_object</span></span>
<span><span class="co">## --------------</span></span>
<span><span class="co">## Dataset key: NA </span></span>
<span><span class="co">## Batch key: Batch </span></span>
<span><span class="co">## Number of datasets: 1 </span></span>
<span><span class="co">## Total number of batches: 4 </span></span>
<span><span class="co">## Batches (showing first 5):</span></span>
<span><span class="co">## Batch1, Batch2, Batch3, Batch4 </span></span>
<span><span class="co">## Number of intersected genes: 4964 </span></span>
<span><span class="co">## Intersected genes (showing first 10):</span></span>
<span><span class="co">## Gene1, Gene2, Gene3, Gene4, Gene5, Gene6, Gene7, Gene8, Gene9, Gene10 , ...</span></span>
<span><span class="co">## Min cells per gene: 1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-anglemania-function">run anglemania function<a class="anchor" aria-label="anchor" href="#run-anglemania-function"></a>
</h3>
<ul>
<li>the anglemania function has five arguments
<ul>
<li>the input anglemania object</li>
<li>method: either cosine, spearman or diem
<ul>
<li>this is the method by which the relationship of the gene pairs is
measured. Default is cosine, which is the cosine similarity between the
expression vectors of the gene pairs.</li>
</ul>
</li>
<li>zscore_mean_threshold: We compute a mean of the zscore of the
relationship between a gene pair, and then we set a minimal cutoff for
the (absolute) mean zscore. A cutoff of 2 means that the filtered gene
pairs have a relationship, e.g. cosine similarity, that is 2 standard
deviations away from the mean of all cosine similarities from this
dataset. A higher value means a more stringent cutoff.</li>
<li>zscore_sn_threshold: The SN or signal-to-noise ratio measures the
invariance of the relationship of the relationship between the gene
pair. A high SN ratio means that the relationship is constant over
multiple batches.</li>
<li>max_n_genes: you can specify a maximum number of extracted genes.
They are sorted by decreasing mean zscore</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">angl</span>,</span>
<span>  zscore_mean_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  zscore_sn_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  max_n_genes <span class="op">=</span> <span class="fl">2000</span> <span class="co"># optionally define a max number of genes.</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing angles and transforming to z-scores...</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmp3u9iaN/file2354617d52f6" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmp3u9iaN/file23541a8c7bb5" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmp3u9iaN/file2354701974f" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmp3u9iaN/file23541d907be7" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Computing statistics...</span></span></code></pre>
<pre><code><span><span class="co">## Weighting matrix_list...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating mean...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating sds...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering features...</span></span></code></pre>
<pre><code><span><span class="co">## Selecting features...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Selected 1496 genes for integration."</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If you think the number of selected genes is too high or low you can adjust the thresholds:</span></span>
<span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemanise_utils.html">select_genes</a></span><span class="op">(</span><span class="va">angl</span>,</span>
<span>                     zscore_mean_threshold <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>                     zscore_sn_threshold <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Selected 306 genes for integration."</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect the anglemania genes</span></span>
<span><span class="va">integration_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_object-methods.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">integration_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene394"  "Gene3527" "Gene71"   "Gene5000" "Gene1055" "Gene4201"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">integration_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 306</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="mnn-integration">MNN integration<a class="anchor" aria-label="anchor" href="#mnn-integration"></a>
</h2>
<div class="section level3">
<h3 id="hvgs">HVGs<a class="anchor" aria-label="anchor" href="#hvgs"></a>
</h3>
<div class="section level4">
<h4 id="hvgs-1">320 HVGs<a class="anchor" aria-label="anchor" href="#hvgs-1"></a>
</h4>
</div>
<div class="section level4">
<h4 id="hvgs-2">2000 HVGs<a class="anchor" aria-label="anchor" href="#hvgs-2"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="anglemania">anglemania<a class="anchor" aria-label="anchor" href="#anglemania"></a>
</h3>
</div>
<div class="section level3">
<h3 id="plot">Plot<a class="anchor" aria-label="anchor" href="#plot"></a>
</h3>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_anglemania"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using anglemania genes, colored by Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_anglemania"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using anglemania genes, colored by Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-5-2.png" width="700"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_320"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 320 HVGs, colored by Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-5-3.png" width="700"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_320"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 320 HVGs, colored by Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-5-4.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_2000"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 2000 HVGs, colored by Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-5-5.png" width="700"></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_2000"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 2000 HVGs, colored by Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_files/figure-html/unnamed-chunk-5-6.png" width="700"></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] batchelor_1.22.0            bluster_1.16.0             </span></span>
<span><span class="co">##  [3] scran_1.34.0                scater_1.34.1              </span></span>
<span><span class="co">##  [5] ggplot2_3.5.1               scuttle_1.16.0             </span></span>
<span><span class="co">##  [7] splatter_1.30.0             SingleCellExperiment_1.28.1</span></span>
<span><span class="co">##  [9] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">## [11] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        </span></span>
<span><span class="co">## [13] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span><span class="co">## [15] BiocGenerics_0.52.0         MatrixGenerics_1.18.1      </span></span>
<span><span class="co">## [17] matrixStats_1.5.0           Seurat_5.2.1               </span></span>
<span><span class="co">## [19] SeuratObject_5.0.2          sp_2.2-0                   </span></span>
<span><span class="co">## [21] dplyr_1.1.4                 anglemania_0.99.1          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22          splines_4.4.3            </span></span>
<span><span class="co">##   [3] later_1.4.1               tibble_3.2.1             </span></span>
<span><span class="co">##   [5] polyclip_1.10-7           fastDummies_1.7.5        </span></span>
<span><span class="co">##   [7] lifecycle_1.0.4           edgeR_4.4.2              </span></span>
<span><span class="co">##   [9] doParallel_1.0.17         globals_0.16.3           </span></span>
<span><span class="co">##  [11] lattice_0.22-6            MASS_7.3-64              </span></span>
<span><span class="co">##  [13] backports_1.5.0           magrittr_2.0.3           </span></span>
<span><span class="co">##  [15] limma_3.62.2              plotly_4.10.4            </span></span>
<span><span class="co">##  [17] sass_0.4.9                rmarkdown_2.29           </span></span>
<span><span class="co">##  [19] jquerylib_0.1.4           yaml_2.3.10              </span></span>
<span><span class="co">##  [21] bigparallelr_0.3.2        metapod_1.14.0           </span></span>
<span><span class="co">##  [23] httpuv_1.6.15             sctransform_0.4.1        </span></span>
<span><span class="co">##  [25] spam_2.11-1               spatstat.sparse_3.1-0    </span></span>
<span><span class="co">##  [27] reticulate_1.42.0         cowplot_1.1.3            </span></span>
<span><span class="co">##  [29] pbapply_1.7-2             RColorBrewer_1.1-3       </span></span>
<span><span class="co">##  [31] ResidualMatrix_1.16.0     abind_1.4-8              </span></span>
<span><span class="co">##  [33] zlibbioc_1.52.0           Rtsne_0.17               </span></span>
<span><span class="co">##  [35] purrr_1.0.4               bigassertr_0.1.6         </span></span>
<span><span class="co">##  [37] GenomeInfoDbData_1.2.13   ggrepel_0.9.6            </span></span>
<span><span class="co">##  [39] irlba_2.3.5.1             listenv_0.9.1            </span></span>
<span><span class="co">##  [41] spatstat.utils_3.1-3      goftest_1.2-3            </span></span>
<span><span class="co">##  [43] RSpectra_0.16-2           dqrng_0.4.1              </span></span>
<span><span class="co">##  [45] spatstat.random_3.3-3     fitdistrplus_1.2-2       </span></span>
<span><span class="co">##  [47] parallelly_1.43.0         DelayedMatrixStats_1.28.1</span></span>
<span><span class="co">##  [49] pkgdown_2.1.1             codetools_0.2-20         </span></span>
<span><span class="co">##  [51] DelayedArray_0.32.0       tidyselect_1.2.1         </span></span>
<span><span class="co">##  [53] UCSC.utils_1.2.0          farver_2.1.2             </span></span>
<span><span class="co">##  [55] viridis_0.6.5             ScaledMatrix_1.14.0      </span></span>
<span><span class="co">##  [57] bigstatsr_1.6.1           spatstat.explore_3.4-2   </span></span>
<span><span class="co">##  [59] flock_0.7                 jsonlite_2.0.0           </span></span>
<span><span class="co">##  [61] BiocNeighbors_2.0.1       progressr_0.15.1         </span></span>
<span><span class="co">##  [63] ggridges_0.5.6            survival_3.8-3           </span></span>
<span><span class="co">##  [65] iterators_1.0.14          systemfonts_1.2.2        </span></span>
<span><span class="co">##  [67] foreach_1.5.2             tools_4.4.3              </span></span>
<span><span class="co">##  [69] ragg_1.3.3                ica_1.0-3                </span></span>
<span><span class="co">##  [71] Rcpp_1.0.14               glue_1.8.0               </span></span>
<span><span class="co">##  [73] gridExtra_2.3             SparseArray_1.6.2        </span></span>
<span><span class="co">##  [75] xfun_0.52                 withr_3.0.2              </span></span>
<span><span class="co">##  [77] fastmap_1.2.0             rsvd_1.0.5               </span></span>
<span><span class="co">##  [79] digest_0.6.37             R6_2.6.1                 </span></span>
<span><span class="co">##  [81] mime_0.13                 textshaping_1.0.0        </span></span>
<span><span class="co">##  [83] colorspace_2.1-1          scattermore_1.2          </span></span>
<span><span class="co">##  [85] tensor_1.5                spatstat.data_3.1-6      </span></span>
<span><span class="co">##  [87] tidyr_1.3.1               generics_0.1.3           </span></span>
<span><span class="co">##  [89] data.table_1.17.0         FNN_1.1.4.1              </span></span>
<span><span class="co">##  [91] httr_1.4.7                htmlwidgets_1.6.4        </span></span>
<span><span class="co">##  [93] S4Arrays_1.6.0            uwot_0.2.3               </span></span>
<span><span class="co">##  [95] pkgconfig_2.0.3           gtable_0.3.6             </span></span>
<span><span class="co">##  [97] lmtest_0.9-40             XVector_0.46.0           </span></span>
<span><span class="co">##  [99] htmltools_0.5.8.1         dotCall64_1.2            </span></span>
<span><span class="co">## [101] scales_1.3.0              png_0.1-8                </span></span>
<span><span class="co">## [103] spatstat.univar_3.1-2     knitr_1.50               </span></span>
<span><span class="co">## [105] reshape2_1.4.4            checkmate_2.3.2          </span></span>
<span><span class="co">## [107] nlme_3.1-167              cachem_1.1.0             </span></span>
<span><span class="co">## [109] zoo_1.8-13                stringr_1.5.1            </span></span>
<span><span class="co">## [111] rmio_0.4.0                KernSmooth_2.23-26       </span></span>
<span><span class="co">## [113] vipor_0.4.7               parallel_4.4.3           </span></span>
<span><span class="co">## [115] miniUI_0.1.1.1            desc_1.4.3               </span></span>
<span><span class="co">## [117] pillar_1.10.2             grid_4.4.3               </span></span>
<span><span class="co">## [119] vctrs_0.6.5               RANN_2.6.2               </span></span>
<span><span class="co">## [121] promises_1.3.2            BiocSingular_1.22.0      </span></span>
<span><span class="co">## [123] ff_4.5.2                  beachmat_2.22.0          </span></span>
<span><span class="co">## [125] xtable_1.8-4              cluster_2.1.8            </span></span>
<span><span class="co">## [127] beeswarm_0.4.0            evaluate_1.0.3           </span></span>
<span><span class="co">## [129] cli_3.6.4                 locfit_1.5-9.12          </span></span>
<span><span class="co">## [131] compiler_4.4.3            rlang_1.1.5              </span></span>
<span><span class="co">## [133] crayon_1.5.3              future.apply_1.11.3      </span></span>
<span><span class="co">## [135] labeling_0.4.3            ps_1.9.0                 </span></span>
<span><span class="co">## [137] ggbeeswarm_0.7.2          plyr_1.8.9               </span></span>
<span><span class="co">## [139] fs_1.6.5                  stringi_1.8.7            </span></span>
<span><span class="co">## [141] viridisLite_0.4.2         deldir_2.0-4             </span></span>
<span><span class="co">## [143] BiocParallel_1.40.1       munsell_0.5.1            </span></span>
<span><span class="co">## [145] lazyeval_0.2.2            spatstat.geom_3.3-6      </span></span>
<span><span class="co">## [147] Matrix_1.7-2              RcppHNSW_0.6.0           </span></span>
<span><span class="co">## [149] patchwork_1.3.0           sparseMatrixStats_1.18.0 </span></span>
<span><span class="co">## [151] future_1.34.0             statmod_1.5.0            </span></span>
<span><span class="co">## [153] shiny_1.10.0              ROCR_1.0-11              </span></span>
<span><span class="co">## [155] igraph_2.1.4              bslib_0.9.0              </span></span>
<span><span class="co">## [157] bit_4.6.0</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Aaron Kollotzek, Vedran Franke, Artem Baranovskii, Altuna Akalin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
