<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>anglemania Tutorial • anglemania</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="anglemania Tutorial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">anglemania</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/anglemania_tutorial.html">anglemania Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>anglemania Tutorial</h1>
                        <h4 data-toc-skip class="author">Aaron
Kollotzek</h4>
            <address class="author_afil">
      Max Delbrück Center for Molecular Medicine in the Helmholtz
Association, Berlin, GermanyBerlin Institute for Medical Systems
Biology, Berlin,
Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:aaron.kollotzek@mdc-berlin.de" class="email">aaron.kollotzek@mdc-berlin.de</a>
      </address>
                              <h4 data-toc-skip class="author">Vedran
Franke</h4>
            <address class="author_afil">
      Max Delbrück Center for Molecular Medicine in the Helmholtz
Association, Berlin, GermanyBerlin Institute for Medical Systems
Biology, Berlin,
Germany<br><h4 data-toc-skip class="author">Artem
Baranovskii</h4>
            <address class="author_afil">
      Helmholtz
Munich<br><h4 data-toc-skip class="author">Altuna
Akalin</h4>
            <address class="author_afil">
      Max Delbrück Center for Molecular Medicine in the Helmholtz
Association, Berlin, GermanyBerlin Institute for Medical Systems
Biology, Berlin, Germany<br><h4 data-toc-skip class="date">2025-09-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BIMSBbioinfo/anglemania/blob/master/vignettes/anglemania_tutorial.Rmd" class="external-link"><code>vignettes/anglemania_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>anglemania_tutorial.Rmd</code></div>

    </address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>anglemania is a feature selection package that extracts genes from
multi-batch scRNA-seq experiments for downstream dataset integration.
The goal is to select genes that carry high biological information and
low technical noise between the batches. Those genes are extracted from
gene pairs that have an invariant and extremely narrow or wide angle
between their expression vectors. Conventionally, highly-variable genes
(HVGs) or sometimes all genes are used for integration tasks (<a href="https://www.nature.com/articles/s41592-021-01336-8" class="external-link uri">https://www.nature.com/articles/s41592-021-01336-8</a>).
While HVGs are a great and easy way to reduce the noise and
dimensionality of the data, we hypothesize that there are better ways to
select genes specifically for integration tasks. HVGs are sensitive to
batch effects because the variance is a function of both the technical
and biological variance. anglemania improves conventional usage of HVGs
for integration tasks, especially when the transcriptional difference
between cell types or cell states is subtle (showcased here using
simulated data generated using <em><a href="https://bioconductor.org/packages/3.21/splatter" class="external-link">splatter</a></em>
with <code>de.facLoc</code> and <code>de.facScale</code> set to 0.1,
which results in mild differences between “Groups”). The package can be
used on top of <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
or <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em>
objects.</p>
<p>Under the hood, anglemania works with file-backed big matrices (FBMs)
from the <em><a href="https://CRAN.R-project.org/package=bigstatsr" class="external-link">bigstatsr</a></em>
package for fast and memory efficient computation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">anglemania</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>Seurat.object.assay.version <span class="op">=</span> <span class="st">"v5"</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/splatter/" class="external-link">splatter</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/hms-dbmi/UpSetR" class="external-link">UpSetR</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by</span></span>
<span><span class="co">## 'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="simulation">Simulation<a class="anchor" aria-label="anchor" href="#simulation"></a>
</h2>
<p>We simulate a scRNA-seq dataset using Splatter with 4 batches and 3
cell types with subtle differences between cell types and rather big
batch effects.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch.facLoc</span> <span class="op">&lt;-</span> <span class="fl">0.3</span></span>
<span><span class="va">de.facLoc</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">nBatches</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="va">nGroups</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span>
<span><span class="va">groupCells</span> <span class="op">&lt;-</span> <span class="fl">300</span></span>
<span></span>
<span><span class="va">sce_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://oshlacklab.com/splatter/reference/splatSimulate.html" class="external-link">splatSimulate</a></span><span class="op">(</span></span>
<span>    batchCells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">groupCells</span> <span class="op">*</span> <span class="va">nGroups</span>, <span class="va">nBatches</span><span class="op">)</span>,</span>
<span>    batch.facLoc <span class="op">=</span> <span class="va">batch.facLoc</span>,</span>
<span>    group.prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="va">nGroups</span>, <span class="va">nGroups</span><span class="op">)</span>,</span>
<span>    nGenes <span class="op">=</span> <span class="va">nGenes</span>,</span>
<span>    batch.facScale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"groups"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    out.prob <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>    de.prob <span class="op">=</span> <span class="fl">0.1</span>, <span class="co"># mild</span></span>
<span>    de.facLoc <span class="op">=</span> <span class="va">de.facLoc</span>,</span>
<span>    de.facScale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    bcv.common <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    seed <span class="op">=</span> <span class="fl">42</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce_raw</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 6</span></span>
<span><span class="co">## names(6): BatchCellMeans BaseCellMeans BCV CellMeans TrueCounts counts</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="unintegrated-data">Unintegrated data<a class="anchor" aria-label="anchor" href="#unintegrated-data"></a>
</h2>
<p>Here we perform a standard workflow on the unintegrated data. When we
perform clustering on the unintegrated data and visualize it in a UMAP,
we can see that the clusters are driven by batch effects rather than
cell types.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="va">sce</span></span>
<span><span class="co"># Normalization.</span></span>
<span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce_unintegrated</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Feature selection.</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce_unintegrated</span><span class="op">)</span></span>
<span><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, prop <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span></span>
<span>    <span class="va">sce_unintegrated</span>,</span>
<span>    ncomponents <span class="op">=</span> <span class="fl">50</span>,</span>
<span>    subset_row <span class="op">=</span> <span class="va">hvg</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Clustering.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce_unintegrated</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>,</span>
<span>    use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualization.</span></span>
<span><span class="va">sce_unintegrated</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Unintegrated data, colored by Batch"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce_unintegrated</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Unintegrated data, colored by Group"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="anglemania_tutorial_files/figure-html/unnamed-chunk-1-1.png" alt="UMAPs of unintegrated data, colored by Batch and Group. The clusters are driven by batch effects." width="960"><p class="caption">
UMAPs of unintegrated data, colored by Batch and Group. The clusters are
driven by batch effects.
</p>
</div>
</div>
<div class="section level2">
<h2 id="anglemania">anglemania<a class="anchor" aria-label="anchor" href="#anglemania"></a>
</h2>
<p>anglemania works on a <em><a href="https://bioconductor.org/packages/3.21/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
object. The function has a few important arguments: -
<code>batch_key</code>: the column in the metadata of the
SingleCellExperiment object that indicates which batch the cells belong
to. This is required to distinguish between batches, because we compute
the angle between gene pairs for each batch. - <code>method</code>:
either cosine, spearman or diem - this is the method by which the
relationship of the gene pairs is measured. Default is cosine, which is
the cosine similarity between the expression vectors of the gene pairs.
- <code>zscore_mean_threshold</code>: We compute a mean of the zscore of
the relationship between a gene pair, and then we set a minimal cutoff
for the (absolute) mean zscore. A cutoff of 2 means that the filtered
gene pairs have a relationship, e.g. cosine similarity, that is 2
standard deviations away from the mean of all cosine similarities from
this dataset. A higher value means a more stringent cutoff. -
<code>zscore_sn_threshold</code>: The SNR or signal-to-noise ratio
measures the invariance of the relationship of the relationship between
the gene pair. A high SN ratio means that the relationship is constant
over multiple batches. - <code>max_n_genes</code>: you can specify a
maximum number of extracted genes. They are sorted by decreasing mean
zscore after passing the thresholds.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">##              Cell       Batch    Group ExpLibSize</span></span>
<span><span class="co">##       &lt;character&gt; &lt;character&gt; &lt;factor&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## Cell1       Cell1      Batch1   Group1    46898.1</span></span>
<span><span class="co">## Cell2       Cell2      Batch1   Group1    54688.2</span></span>
<span><span class="co">## Cell3       Cell3      Batch1   Group2    52027.9</span></span>
<span><span class="co">## Cell4       Cell4      Batch1   Group1    52319.5</span></span>
<span><span class="co">## Cell5       Cell5      Batch1   Group3    37774.7</span></span>
<span><span class="co">## Cell6       Cell6      Batch1   Group3    58112.3</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch_key</span> <span class="op">&lt;-</span> <span class="st">"Batch"</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania.html">anglemania</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    batch_key <span class="op">=</span> <span class="va">batch_key</span>,</span>
<span>    zscore_mean_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    zscore_sn_threshold <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="extract-the-anglemania-genes-from-the-sce-object">extract the anglemania genes from the SCE object<a class="anchor" aria-label="anchor" href="#extract-the-anglemania-genes-from-the-sce-object"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anglemania_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_utils.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anglemania_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene394"  "Gene3527" "Gene71"   "Gene5000" "Gene1055" "Gene4201"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">anglemania_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1496</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="select_genes">select_genes<a class="anchor" aria-label="anchor" href="#select_genes"></a>
</h3>
<p>once anglemania was run on the SCE, you can adjust the initial zscore
mean and zscore SNR thresholds by using the <code><a href="../reference/select_genes.html">select_genes()</a></code>
function</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># If you think the number of selected genes is</span></span>
<span><span class="co"># too high or low you can adjust the thresholds:</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_genes.html">select_genes</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    zscore_mean_threshold <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>    zscore_sn_threshold <span class="op">=</span> <span class="fl">2.5</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Inspect the anglemania genes</span></span>
<span><span class="va">anglemania_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_utils.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anglemania_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene394"  "Gene3527" "Gene71"   "Gene5000" "Gene1055" "Gene4201"</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">anglemania_genes</span><span class="op">)</span> <span class="co"># 306 genes are selected with these thresholds</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 306</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="mnn-integration">MNN integration<a class="anchor" aria-label="anchor" href="#mnn-integration"></a>
</h2>
<p>The anglemania genes can now be used for downstream integration
algorithms such as MNN. We compare the integration results using the
anglemania genes with the results using 300 and the standard 2000 HVGs.
## HVGs ### 300 HVGs</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvg_300</span> <span class="op">&lt;-</span> <span class="va">sce</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span>block <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="va">barcodes_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sce_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">barcodes_by_batch</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">x</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sce_mnn</span> <span class="op">&lt;-</span> <span class="va">sce</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce_mnn</span> <span class="op">&lt;-</span> <span class="fu">batchelor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">fastMNN</a></span><span class="op">(</span></span>
<span>    <span class="va">sce_mnn</span>,</span>
<span>    subset.row <span class="op">=</span> <span class="va">hvg_300</span>,</span>
<span>    k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    batch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_mnn</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    d <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"MNN_hvg_300"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce_mnn</span>, <span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"MNN_hvg_300"</span>, name <span class="op">=</span> <span class="st">"umap_MNN_hvg_300"</span><span class="op">)</span></span>
<span><span class="co"># k is the number of nearest neighbours to consider when identifying MNNs</span></span></code></pre></div>
<div class="section level4">
<h4 id="hvgs">2000 HVGs<a class="anchor" aria-label="anchor" href="#hvgs"></a>
</h4>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvg_2000</span> <span class="op">&lt;-</span> <span class="va">sce</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span>block <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="va">barcodes_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sce_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">barcodes_by_batch</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">sce</span><span class="op">[</span>, <span class="va">x</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sce_mnn</span> <span class="op">&lt;-</span> <span class="va">sce</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce_mnn</span> <span class="op">&lt;-</span> <span class="fu">batchelor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">fastMNN</a></span><span class="op">(</span></span>
<span>    <span class="va">sce_mnn</span>,</span>
<span>    subset.row <span class="op">=</span> <span class="va">hvg_2000</span>,</span>
<span>    k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    batch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_mnn</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    d <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"MNN_hvg_2000"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce_mnn</span>, <span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"MNN_hvg_2000"</span>, name <span class="op">=</span> <span class="st">"umap_MNN_hvg_2000"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="anglemania-genes">anglemania genes<a class="anchor" aria-label="anchor" href="#anglemania-genes"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_mnn</span> <span class="op">&lt;-</span> <span class="va">sce</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce_mnn</span> <span class="op">&lt;-</span> <span class="fu">batchelor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">fastMNN</a></span><span class="op">(</span></span>
<span>    <span class="va">sce_mnn</span>,</span>
<span>    subset.row <span class="op">=</span> <span class="va">anglemania_genes</span>,</span>
<span>    k <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    batch <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_mnn</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    d <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"MNN_anglemania"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce_mnn</span>, <span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    dimred <span class="op">=</span> <span class="st">"MNN_anglemania"</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"umap_MNN_anglemania"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot">Plot<a class="anchor" aria-label="anchor" href="#plot"></a>
</h3>
<div class="section level4">
<h4 id="umap-embeddings">UMAP embeddings<a class="anchor" aria-label="anchor" href="#umap-embeddings"></a>
</h4>
<p>We can see from the UMAPs that anglemania genes yield the best
integration in terms of clustering by cell type and mixing the batches.
The goal of an integration and subsequent clustering should be to have
low intra cluster variance and high inter cluster variance. This is at
least true for most downstream scRNA-seq analyses where the goal is to
e.g. differentiate between cell types or cell states and annotate
these.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use wrap_plots</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_anglemania"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using anglemania genes, colored by Batch"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_anglemania"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using anglemania genes, colored by Group"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_300"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 300 HVGs, colored by Batch"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_300"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 300 HVGs, colored by Group"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Batch"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_2000"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 2000 HVGs, colored by Batch"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"Group"</span>, dimred <span class="op">=</span> <span class="st">"umap_MNN_hvg_2000"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MNN integration using top 2000 HVGs, colored by Group"</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="anglemania_tutorial_files/figure-html/unnamed-chunk-3-1.png" alt="UMAPs of MNN integrated data. Comparison of UMAP embeddings of integrated data using anglemania genes, top 300 HVGs and top 2000 HVGs." width="1152"><p class="caption">
UMAPs of MNN integrated data. Comparison of UMAP embeddings of
integrated data using anglemania genes, top 300 HVGs and top 2000 HVGs.
</p>
</div>
</div>
<div class="section level4">
<h4 id="overlap">Overlap<a class="anchor" aria-label="anchor" href="#overlap"></a>
</h4>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">upsetr_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html" class="external-link">fromList</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        anglemania <span class="op">=</span> <span class="va">anglemania_genes</span>,</span>
<span>        hvg_300 <span class="op">=</span> <span class="va">hvg_300</span>,</span>
<span>        hvg_2000 <span class="op">=</span> <span class="va">hvg_2000</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html" class="external-link">upset</a></span><span class="op">(</span><span class="va">upsetr_df</span>, text.scale <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="anglemania_tutorial_files/figure-html/unnamed-chunk-4-1.png" alt="Overlap of selected genes. Additionally, we check the overlap of the anglemania genes with the HVGs. About 33 of the 306 anglemania genes are also found in the top 300 HVGs, and about 179 of the 306 anglemania genes are also found in the top 2000 HVGs." width="960"><p class="caption">
Overlap of selected genes. Additionally, we check the overlap of the
anglemania genes with the HVGs. About 33 of the 306 anglemania genes are
also found in the top 300 HVGs, and about 179 of the 306 anglemania
genes are also found in the top 2000 HVGs.
</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a>
</h2>
<p>Now you can just use the anglemania genes for other integration
algorithms. When using Seurat, the easiest approach is to create an SCE
from the counts and metadata of the SeuratObject, then run anglemania on
it and save those genes as the VariableFeatures of the SeuratObject.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce_raw</span><span class="op">)</span>,</span>
<span>    meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce_raw</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 5000 features across 3600 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (5000 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anglemania_genes</span> <span class="op">&lt;-</span> <span class="va">se</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span>assay <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/anglemania.html">anglemania</a></span><span class="op">(</span></span>
<span>        batch_key <span class="op">=</span> <span class="st">"Batch"</span>,</span>
<span>        zscore_mean_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        zscore_sn_threshold <span class="op">=</span> <span class="fl">2</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/anglemania_utils.html">get_anglemania_genes</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="integration">Integration<a class="anchor" aria-label="anchor" href="#integration"></a>
</h3>
<p>In Seurat v5 you split the layers of an assay by batch and then run
the normal Seurat workflow. To use anglemania genes for integration, you
need to assign them to the VariableFeatures slot of the SeuratObject.
After that, you integrate the layers using the anglemania genes as the
<code>features</code> argument.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Split by batch</span></span>
<span><span class="va">se</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">se</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">se</span><span class="op">$</span><span class="va">Batch</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Standard preprocessing but use anglemania genes as "VariableFeatures"</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">se</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">anglemania_genes</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Integrate</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">se</span>,</span>
<span>    method <span class="op">=</span> <span class="va">CCAIntegration</span>,</span>
<span>    orig.reduction <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>    new.reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>,</span>
<span>    features <span class="op">=</span> <span class="va">anglemania_genes</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">se</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction <span class="op">=</span> <span class="st">"integrated.cca"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 5000 features across 3600 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (5000 features, 1496 variable features)</span></span>
<span><span class="co">##  9 layers present: counts.Batch1, counts.Batch2, counts.Batch3, counts.Batch4, data.Batch1, data.Batch2, data.Batch3, data.Batch4, scale.data</span></span>
<span><span class="co">##  3 dimensional reductions calculated: pca, integrated.cca, umap</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="plot-1">Plot<a class="anchor" aria-label="anchor" href="#plot-1"></a>
</h3>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">se</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Seurat integration using\nanglemania genes\ncolored by Batch"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">se</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Seurat integration using\nanglemania genes\ncolored by Group"</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="anglemania_tutorial_files/figure-html/unnamed-chunk-6-1.png" alt="UMAPs of Seurat integrated data. Here we show that we can use the anglemania genes for integration of a SeuratObject." width="960"><p class="caption">
UMAPs of Seurat integrated data. Here we show that we can use the
anglemania genes for integration of a SeuratObject.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="showcase-underlying-functions">Showcase underlying functions<a class="anchor" aria-label="anchor" href="#showcase-underlying-functions"></a>
</h2>
<div class="section level3">
<h3 id="normal-anglemania-workflow">Normal anglemania workflow<a class="anchor" aria-label="anchor" href="#normal-anglemania-workflow"></a>
</h3>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce_raw</span></span>
<span><span class="va">batch_key</span> <span class="op">&lt;-</span> <span class="st">"batch"</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="va">batch_key</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><code>anglemania</code> is run on the SCE object and it basically
calls three functions:</p>
<ul>
<li>
<code>factorise</code>:
<ul>
<li>creates a permutation of the input matrix whose correlation matrix
is used to create a null distribution for each batch.</li>
<li>computes the cosine similarity (or spearman coefficient) between
gene expression vector pairs matrix for both the original and permuted
matrices</li>
<li>computes the zscore of the relationship between the gene pairs
taking the mean and standard deviation of the null distribution</li>
<li>it does this for every batch in the dataset!</li>
</ul>
</li>
<li>
<code>get_list_stats</code>
<ul>
<li>computes the mean and standard deviation of the zscores across the
matrices from the different batches. This creates two important
matrices: the mean zscore matrix <code>mean_zscore</code> and the
signal-to-noise ratio matrix <code>sn_zscore</code>. These are stored in
the metadata of the SCE object.</li>
</ul>
</li>
<li>
<code>select_genes</code>
<ul>
<li>filters the gene pairs by the <code>mean_zscore</code> and
<code>sn_zscore</code> matrices (SN ratio, i.e. the mean divided by the
standard deviation).</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="factorise">factorise<a class="anchor" aria-label="anchor" href="#factorise"></a>
</h3>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barcodes_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">counts_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">barcodes_by_batch</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[</span>, <span class="va">x</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/anglemania_utils.html">sparse_to_fbm</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">counts_by_batch</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       [,1] [,2] [,3] [,4] [,5] [,6]</span></span>
<span><span class="co">##  [1,]    8    5    5    2    4    2</span></span>
<span><span class="co">##  [2,]    9    5    4    1    4    9</span></span>
<span><span class="co">##  [3,]    4    2    7    4    6    1</span></span>
<span><span class="co">##  [4,]    7    4    4    1    3    8</span></span>
<span><span class="co">##  [5,]    6    8    5    5    7    4</span></span>
<span><span class="co">##  [6,]    5    8    5    9    8    6</span></span>
<span><span class="co">##  [7,]    6    4    7    4    7    5</span></span>
<span><span class="co">##  [8,]    3    3    3    3    5    4</span></span>
<span><span class="co">##  [9,]    6    4    5    5    3    1</span></span>
<span><span class="co">## [10,]    6    4    5    2    6    4</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we are working on FBMs (file-backed matrices</span></span>
<span><span class="co"># implemented in the bigstatsr package)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">counts_by_batch</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "FBM"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "bigstatsr"</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># factorise produces the correlation matrices transformed to z-scores</span></span>
<span><span class="va">factorised</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">counts_by_batch</span>, <span class="va">factorise</span><span class="op">)</span></span>
<span><span class="va">factorised</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              [,1]        [,2]        [,3]       [,4]       [,5]        [,6]</span></span>
<span><span class="co">##  [1,]  0.00000000 -1.74668209 -0.79613304  1.1776966  0.8601212  0.50193905</span></span>
<span><span class="co">##  [2,] -1.73785984  0.00000000  0.59661885  2.5115045  0.5348975  1.43483620</span></span>
<span><span class="co">##  [3,] -0.85963025  0.53988067  0.00000000  0.2901283 -1.5767324 -0.06260101</span></span>
<span><span class="co">##  [4,]  1.13658651  2.45514150  0.33158301  0.0000000 -0.3715653 -0.57325460</span></span>
<span><span class="co">##  [5,]  0.92122274  0.57796947 -1.62420449 -0.3961608  0.0000000 -0.26750572</span></span>
<span><span class="co">##  [6,]  0.52925575  1.57028828 -0.02415793 -0.6582967 -0.2979901  0.00000000</span></span>
<span><span class="co">##  [7,] -0.41325531 -0.10548957 -1.16180200  0.0055108  0.3444741  1.41598180</span></span>
<span><span class="co">##  [8,]  0.10001406 -1.38192310 -1.29577006 -0.2524730 -0.7968541 -1.56575595</span></span>
<span><span class="co">##  [9,]  0.51996703  0.05963883 -0.37002326 -0.2927571 -0.9185203 -0.12159157</span></span>
<span><span class="co">## [10,]  0.06970488 -0.33637803  1.57056534  0.3728527  0.5142178  1.01784989</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="get_list_stats">get_list_stats<a class="anchor" aria-label="anchor" href="#get_list_stats"></a>
</h3>
<p>The “list stats” are computed by <code>get_list_stats</code> and take
the z-score transformed correlation matrices from <code>factorise</code>
as input. The outputs are the mean zscore matrix
<code>mean_zscore</code> and the signal-to-noise ratio matrix
<code>sn_zscore</code>. These are stored in the metadata of the SCE
object.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">matrix_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">anglemania</span><span class="op">$</span><span class="va">matrix_list</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">anglemania</span><span class="op">$</span><span class="va">weights</span><span class="op">$</span><span class="va">weight</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">anglemania</span><span class="op">$</span><span class="va">weights</span><span class="op">$</span><span class="va">batch</span></span>
<span><span class="op">)</span></span>
<span><span class="va">list_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/statistics.html">get_list_stats</a></span><span class="op">(</span></span>
<span>    matrix_list <span class="op">=</span> <span class="va">matrix_list</span>,</span>
<span>    weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">list_stats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "mean_zscore" "sds_zscore"  "sn_zscore"</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">list_stats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">list_stats</span><span class="op">$</span><span class="va">mean_zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              [,1]       [,2]       [,3]        [,4]        [,5]       [,6]</span></span>
<span><span class="co">##  [1,]  0.00000000 -1.4773877 -0.6690115  0.80957395  0.21191114  0.3065459</span></span>
<span><span class="co">##  [2,] -1.43297508  0.0000000  0.8094372  1.79591873  0.87491750  0.0903189</span></span>
<span><span class="co">##  [3,] -0.70975605  0.8090174  0.0000000 -0.90002758 -0.14215190  0.4217798</span></span>
<span><span class="co">##  [4,]  0.81025807  1.7916458 -0.8103351  0.00000000 -0.29386285 -0.2877692</span></span>
<span><span class="co">##  [5,]  0.26228504  0.8796571 -0.2190412 -0.31429554  0.00000000 -0.7944232</span></span>
<span><span class="co">##  [6,]  0.34210706  0.1727196  0.4235970 -0.32724467 -0.79625202  0.0000000</span></span>
<span><span class="co">##  [7,]  0.60813294  0.4576137 -1.4403958 -0.03807431  0.35694728  0.5394522</span></span>
<span><span class="co">##  [8,] -0.03868058 -1.0958225 -0.7680151 -0.62207485 -1.42724690 -0.8197054</span></span>
<span><span class="co">##  [9,]  1.26715940  0.3486664 -0.5030029 -0.76152650 -0.66668260 -0.4825756</span></span>
<span><span class="co">## [10,]  0.49115807  1.1467226  0.7976727 -0.27013900  0.03028304  0.8112288</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">list_stats</span><span class="op">$</span><span class="va">sn_zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##            [,1]       [,2]      [,3]      [,4]        [,5]      [,6]</span></span>
<span><span class="co">##  [1,]        NA 3.87928890 3.7213414 1.5550667  0.23116551 1.1093569</span></span>
<span><span class="co">##  [2,] 3.3234406         NA 2.6894229 1.7746389  1.81948152 0.0475004</span></span>
<span><span class="co">##  [3,] 3.3486305 2.12554305        NA 0.5347330  0.07006687 0.6157208</span></span>
<span><span class="co">##  [4,] 1.7557127 1.90940919 0.5017815        NA  2.67420836 0.7127632</span></span>
<span><span class="co">##  [5,] 0.2814584 2.06177351 0.1102260 2.7147101          NA 1.0660911</span></span>
<span><span class="co">##  [6,] 1.2925884 0.08738832 0.6689559 0.6989744  1.12999842        NA</span></span>
<span><span class="co">##  [7,] 0.4210103 0.57464016 3.6559091 0.6177018 20.23533063 0.4351825</span></span>
<span><span class="co">##  [8,] 0.1972052 2.70836022 1.0290168 1.1901274  1.60093202 0.7769169</span></span>
<span><span class="co">##  [9,] 1.1991785 0.85301337 2.6746701 1.1487110  1.87190277 0.9452842</span></span>
<span><span class="co">## [10,] 0.8240564 0.54672981 0.7297777 0.2970755  0.04424840 2.7762187</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Or we can access them directly from the SCE object</span></span>
<span><span class="co"># after running anglemania</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">anglemania</span><span class="op">$</span><span class="va">list_stats</span><span class="op">$</span><span class="va">mean_zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              [,1]       [,2]       [,3]        [,4]        [,5]       [,6]</span></span>
<span><span class="co">##  [1,]  0.00000000 -1.4773877 -0.6690115  0.80957395  0.21191114  0.3065459</span></span>
<span><span class="co">##  [2,] -1.43297508  0.0000000  0.8094372  1.79591873  0.87491750  0.0903189</span></span>
<span><span class="co">##  [3,] -0.70975605  0.8090174  0.0000000 -0.90002758 -0.14215190  0.4217798</span></span>
<span><span class="co">##  [4,]  0.81025807  1.7916458 -0.8103351  0.00000000 -0.29386285 -0.2877692</span></span>
<span><span class="co">##  [5,]  0.26228504  0.8796571 -0.2190412 -0.31429554  0.00000000 -0.7944232</span></span>
<span><span class="co">##  [6,]  0.34210706  0.1727196  0.4235970 -0.32724467 -0.79625202  0.0000000</span></span>
<span><span class="co">##  [7,]  0.60813294  0.4576137 -1.4403958 -0.03807431  0.35694728  0.5394522</span></span>
<span><span class="co">##  [8,] -0.03868058 -1.0958225 -0.7680151 -0.62207485 -1.42724690 -0.8197054</span></span>
<span><span class="co">##  [9,]  1.26715940  0.3486664 -0.5030029 -0.76152650 -0.66668260 -0.4825756</span></span>
<span><span class="co">## [10,]  0.49115807  1.1467226  0.7976727 -0.27013900  0.03028304  0.8112288</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">anglemania</span><span class="op">$</span><span class="va">list_stats</span><span class="op">$</span><span class="va">sn_zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##            [,1]       [,2]      [,3]      [,4]        [,5]      [,6]</span></span>
<span><span class="co">##  [1,]        NA 3.87928890 3.7213414 1.5550667  0.23116551 1.1093569</span></span>
<span><span class="co">##  [2,] 3.3234406         NA 2.6894229 1.7746389  1.81948152 0.0475004</span></span>
<span><span class="co">##  [3,] 3.3486305 2.12554305        NA 0.5347330  0.07006687 0.6157208</span></span>
<span><span class="co">##  [4,] 1.7557127 1.90940919 0.5017815        NA  2.67420836 0.7127632</span></span>
<span><span class="co">##  [5,] 0.2814584 2.06177351 0.1102260 2.7147101          NA 1.0660911</span></span>
<span><span class="co">##  [6,] 1.2925884 0.08738832 0.6689559 0.6989744  1.12999842        NA</span></span>
<span><span class="co">##  [7,] 0.4210103 0.57464016 3.6559091 0.6177018 20.23533063 0.4351825</span></span>
<span><span class="co">##  [8,] 0.1972052 2.70836022 1.0290168 1.1901274  1.60093202 0.7769169</span></span>
<span><span class="co">##  [9,] 1.1991785 0.85301337 2.6746701 1.1487110  1.87190277 0.9452842</span></span>
<span><span class="co">## [10,] 0.8240564 0.54672981 0.7297777 0.2970755  0.04424840 2.7762187</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="select_genes-1">select_genes<a class="anchor" aria-label="anchor" href="#select_genes-1"></a>
</h3>
<ul>
<li>under the hood, <code>anglemania</code> calls
<code>select_genes</code> with the default thresholds
<code>zscore_mean_threshold = 2.5</code>,
<code>zscore_sn_threshold = 2.5</code>
</li>
<li>we can use <code>select_genes</code> to change the thresholds
without having to run anglemania again</li>
</ul>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">previous_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_utils.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_genes.html">select_genes</a></span><span class="op">(</span></span>
<span>    <span class="va">sce</span>,</span>
<span>    zscore_mean_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    zscore_sn_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Inspect the anglemania genes</span></span>
<span><span class="va">new_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_utils.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">previous_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 25</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">new_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 194</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] future_1.67.0               UpSetR_1.4.0               </span></span>
<span><span class="co">##  [3] batchelor_1.24.0            bluster_1.18.0             </span></span>
<span><span class="co">##  [5] scran_1.36.0                scater_1.36.0              </span></span>
<span><span class="co">##  [7] ggplot2_4.0.0               scuttle_1.18.0             </span></span>
<span><span class="co">##  [9] splatter_1.32.0             SingleCellExperiment_1.30.1</span></span>
<span><span class="co">## [11] SummarizedExperiment_1.38.1 Biobase_2.68.0             </span></span>
<span><span class="co">## [13] GenomicRanges_1.60.0        GenomeInfoDb_1.44.3        </span></span>
<span><span class="co">## [15] IRanges_2.42.0              S4Vectors_0.46.0           </span></span>
<span><span class="co">## [17] BiocGenerics_0.54.0         generics_0.1.4             </span></span>
<span><span class="co">## [19] MatrixGenerics_1.20.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [21] Seurat_5.3.0                SeuratObject_5.2.0         </span></span>
<span><span class="co">## [23] sp_2.2-0                    dplyr_1.1.4                </span></span>
<span><span class="co">## [25] anglemania_0.99.5           BiocStyle_2.36.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22          splines_4.5.1            </span></span>
<span><span class="co">##   [3] later_1.4.4               tibble_3.3.0             </span></span>
<span><span class="co">##   [5] polyclip_1.10-7           fastDummies_1.7.5        </span></span>
<span><span class="co">##   [7] lifecycle_1.0.4           edgeR_4.6.3              </span></span>
<span><span class="co">##   [9] doParallel_1.0.17         globals_0.18.0           </span></span>
<span><span class="co">##  [11] lattice_0.22-7            MASS_7.3-65              </span></span>
<span><span class="co">##  [13] backports_1.5.0           magrittr_2.0.4           </span></span>
<span><span class="co">##  [15] limma_3.64.3              plotly_4.11.0            </span></span>
<span><span class="co">##  [17] sass_0.4.10               rmarkdown_2.29           </span></span>
<span><span class="co">##  [19] jquerylib_0.1.4           yaml_2.3.10              </span></span>
<span><span class="co">##  [21] bigparallelr_0.3.2        metapod_1.16.0           </span></span>
<span><span class="co">##  [23] httpuv_1.6.16             sctransform_0.4.2        </span></span>
<span><span class="co">##  [25] spam_2.11-1               spatstat.sparse_3.1-0    </span></span>
<span><span class="co">##  [27] reticulate_1.43.0         cowplot_1.2.0            </span></span>
<span><span class="co">##  [29] pbapply_1.7-4             RColorBrewer_1.1-3       </span></span>
<span><span class="co">##  [31] ResidualMatrix_1.18.0     abind_1.4-8              </span></span>
<span><span class="co">##  [33] Rtsne_0.17                purrr_1.1.0              </span></span>
<span><span class="co">##  [35] bigassertr_0.1.7          GenomeInfoDbData_1.2.14  </span></span>
<span><span class="co">##  [37] ggrepel_0.9.6             irlba_2.3.5.1            </span></span>
<span><span class="co">##  [39] listenv_0.9.1             spatstat.utils_3.2-0     </span></span>
<span><span class="co">##  [41] goftest_1.2-3             RSpectra_0.16-2          </span></span>
<span><span class="co">##  [43] dqrng_0.4.1               spatstat.random_3.4-2    </span></span>
<span><span class="co">##  [45] fitdistrplus_1.2-4        parallelly_1.45.1        </span></span>
<span><span class="co">##  [47] DelayedMatrixStats_1.30.0 pkgdown_2.1.3            </span></span>
<span><span class="co">##  [49] codetools_0.2-20          DelayedArray_0.34.1      </span></span>
<span><span class="co">##  [51] tidyselect_1.2.1          UCSC.utils_1.4.0         </span></span>
<span><span class="co">##  [53] farver_2.1.2              viridis_0.6.5            </span></span>
<span><span class="co">##  [55] ScaledMatrix_1.16.0       bigstatsr_1.6.2          </span></span>
<span><span class="co">##  [57] spatstat.explore_3.5-3    flock_0.7                </span></span>
<span><span class="co">##  [59] jsonlite_2.0.0            BiocNeighbors_2.2.0      </span></span>
<span><span class="co">##  [61] progressr_0.16.0          ggridges_0.5.7           </span></span>
<span><span class="co">##  [63] survival_3.8-3            iterators_1.0.14         </span></span>
<span><span class="co">##  [65] systemfonts_1.2.3         foreach_1.5.2            </span></span>
<span><span class="co">##  [67] tools_4.5.1               ragg_1.5.0               </span></span>
<span><span class="co">##  [69] ica_1.0-3                 Rcpp_1.1.0               </span></span>
<span><span class="co">##  [71] glue_1.8.0                gridExtra_2.3            </span></span>
<span><span class="co">##  [73] SparseArray_1.8.1         xfun_0.53                </span></span>
<span><span class="co">##  [75] withr_3.0.2               BiocManager_1.30.26      </span></span>
<span><span class="co">##  [77] fastmap_1.2.0             rsvd_1.0.5               </span></span>
<span><span class="co">##  [79] digest_0.6.37             R6_2.6.1                 </span></span>
<span><span class="co">##  [81] mime_0.13                 textshaping_1.0.3        </span></span>
<span><span class="co">##  [83] scattermore_1.2           tensor_1.5.1             </span></span>
<span><span class="co">##  [85] spatstat.data_3.1-8       tidyr_1.3.1              </span></span>
<span><span class="co">##  [87] data.table_1.17.8         FNN_1.1.4.1              </span></span>
<span><span class="co">##  [89] httr_1.4.7                htmlwidgets_1.6.4        </span></span>
<span><span class="co">##  [91] S4Arrays_1.8.1            uwot_0.2.3               </span></span>
<span><span class="co">##  [93] pkgconfig_2.0.3           gtable_0.3.6             </span></span>
<span><span class="co">##  [95] lmtest_0.9-40             S7_0.2.0                 </span></span>
<span><span class="co">##  [97] XVector_0.48.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">##  [99] dotCall64_1.2             bookdown_0.44            </span></span>
<span><span class="co">## [101] scales_1.4.0              png_0.1-8                </span></span>
<span><span class="co">## [103] spatstat.univar_3.1-4     knitr_1.50               </span></span>
<span><span class="co">## [105] reshape2_1.4.4            checkmate_2.3.3          </span></span>
<span><span class="co">## [107] nlme_3.1-168              cachem_1.1.0             </span></span>
<span><span class="co">## [109] zoo_1.8-14                stringr_1.5.2            </span></span>
<span><span class="co">## [111] rmio_0.4.0                KernSmooth_2.23-26       </span></span>
<span><span class="co">## [113] vipor_0.4.7               parallel_4.5.1           </span></span>
<span><span class="co">## [115] miniUI_0.1.2              desc_1.4.3               </span></span>
<span><span class="co">## [117] pillar_1.11.1             grid_4.5.1               </span></span>
<span><span class="co">## [119] vctrs_0.6.5               RANN_2.6.2               </span></span>
<span><span class="co">## [121] promises_1.3.3            BiocSingular_1.24.0      </span></span>
<span><span class="co">## [123] ff_4.5.2                  beachmat_2.24.0          </span></span>
<span><span class="co">## [125] xtable_1.8-4              cluster_2.1.8.1          </span></span>
<span><span class="co">## [127] beeswarm_0.4.0            evaluate_1.0.5           </span></span>
<span><span class="co">## [129] cli_3.6.5                 locfit_1.5-9.12          </span></span>
<span><span class="co">## [131] compiler_4.5.1            rlang_1.1.6              </span></span>
<span><span class="co">## [133] crayon_1.5.3              future.apply_1.20.0      </span></span>
<span><span class="co">## [135] labeling_0.4.3            ps_1.9.1                 </span></span>
<span><span class="co">## [137] ggbeeswarm_0.7.2          plyr_1.8.9               </span></span>
<span><span class="co">## [139] fs_1.6.6                  stringi_1.8.7            </span></span>
<span><span class="co">## [141] viridisLite_0.4.2         deldir_2.0-4             </span></span>
<span><span class="co">## [143] BiocParallel_1.42.2       lazyeval_0.2.2           </span></span>
<span><span class="co">## [145] spatstat.geom_3.6-0       Matrix_1.7-3             </span></span>
<span><span class="co">## [147] RcppHNSW_0.6.0            patchwork_1.3.2          </span></span>
<span><span class="co">## [149] sparseMatrixStats_1.20.0  statmod_1.5.0            </span></span>
<span><span class="co">## [151] shiny_1.11.1              ROCR_1.0-11              </span></span>
<span><span class="co">## [153] igraph_2.1.4              bslib_0.9.0              </span></span>
<span><span class="co">## [155] bit_4.6.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Aaron Kollotzek, Vedran Franke, Artem Baranovskii, Altuna Akalin, SFB1588.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
