<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Running anglemania with Seurat • anglemania</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Running anglemania with Seurat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">anglemania</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/anglemania_tutorial_Seurat.html">Running anglemania with Seurat</a>
    </li>
    <li>
      <a href="../articles/anglemania_tutorial.html">Running anglemania with SingleCellExperiment</a>
    </li>
    <li>
      <a href="../articles/Functions_Showcasing.html">Functions_Showcasing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Running anglemania with Seurat</h1>
            
            <h4 data-toc-skip class="date">2025-04-16</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BIMSBbioinfo/anglemania/blob/master/vignettes/anglemania_tutorial_Seurat.Rmd" class="external-link"><code>vignettes/anglemania_tutorial_Seurat.Rmd</code></a></small>
      <div class="hidden name"><code>anglemania_tutorial_Seurat.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>anglemania is a feature selection package that extracts genes from
multi-batch scRNA-seq experiments for downstream dataset integration.
The goal is to select genes that carry high biological information and
low technical noise. Those genes are extracted from gene pairs that have
an invariant and extremely narrow or wide angle between their expression
vectors. It improves conventional usage of highly-variable genes for
integration tasks, especially when the transcriptional difference
between cell types or cell states is subtle (showcased here with
de.facLoc set to 0.1 (= mild differences between “Groups”)). anglemania
can be used on top of SingleCellExperiment or Seurat objects.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinformatics.mdc-berlin.de/anglemania" class="external-link">anglemania</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/splatter/" class="external-link">splatter</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-simulated-data">Create simulated data<a class="anchor" aria-label="anchor" href="#create-simulated-data"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch.facLoc</span> <span class="op">&lt;-</span> <span class="fl">0.3</span></span>
<span><span class="va">de.facLoc</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">nBatches</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span><span class="va">nGroups</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="va">nGenes</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span>
<span><span class="va">groupCells</span> <span class="op">&lt;-</span> <span class="fl">300</span></span>
<span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://oshlack.github.io/splatter/reference/splatSimulate.html" class="external-link">splatSimulate</a></span><span class="op">(</span></span>
<span>    batchCells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">groupCells</span> <span class="op">*</span> <span class="va">nGroups</span>, <span class="va">nBatches</span><span class="op">)</span>,</span>
<span>    batch.facLoc <span class="op">=</span> <span class="va">batch.facLoc</span>,</span>
<span>    group.prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="va">nGroups</span>, <span class="va">nGroups</span><span class="op">)</span>,</span>
<span>    nGenes <span class="op">=</span> <span class="va">nGenes</span>,</span>
<span>    batch.facScale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"groups"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    out.prob    <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>    de.prob     <span class="op">=</span> <span class="fl">0.1</span>, <span class="co"># mild </span></span>
<span>    de.facLoc   <span class="op">=</span> <span class="va">de.facLoc</span>,</span>
<span>    de.facScale <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    bcv.common  <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    seed <span class="op">=</span> <span class="fl">42</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span>, meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sim</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Data is of class matrix. Coercing to dgCMatrix.</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rm(sim)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 5000 features across 3600 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (5000 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Layers.html" class="external-link">Layers</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "counts"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Assays</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "RNA"</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">se</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          orig.ident nCount_RNA nFeature_RNA  Cell  Batch  Group ExpLibSize</span></span>
<span><span class="co">## Cell1 SeuratProject      47324         3707 Cell1 Batch1 Group1   46898.14</span></span>
<span><span class="co">## Cell2 SeuratProject      54908         3850 Cell2 Batch1 Group1   54688.18</span></span>
<span><span class="co">## Cell3 SeuratProject      51501         3799 Cell3 Batch1 Group2   52027.93</span></span>
<span><span class="co">## Cell4 SeuratProject      52272         3862 Cell4 Batch1 Group1   52319.47</span></span>
<span><span class="co">## Cell5 SeuratProject      37808         3569 Cell5 Batch1 Group3   37774.72</span></span>
<span><span class="co">## Cell6 SeuratProject      58801         3843 Cell6 Batch1 Group3   58112.35</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="unintegrated-data">unintegrated data<a class="anchor" aria-label="anchor" href="#unintegrated-data"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_unintegrated</span> <span class="op">&lt;-</span> <span class="va">se</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">se_unintegrated</span> <span class="op">&lt;-</span> <span class="va">se_unintegrated</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Normalizing layer: counts</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<pre><code><span><span class="co">## Centering and scaling data matrix</span></span></code></pre>
<pre><code><span><span class="co">## PC_ 1 </span></span>
<span><span class="co">## Positive:  Gene1367, Gene2172, Gene1089, Gene2598, Gene4513, Gene587, Gene789, Gene1951, Gene503, Gene898 </span></span>
<span><span class="co">##     Gene75, Gene2149, Gene2864, Gene217, Gene394, Gene2276, Gene2803, Gene1839, Gene1716, Gene922 </span></span>
<span><span class="co">##     Gene1226, Gene1075, Gene821, Gene4796, Gene3864, Gene3930, Gene1128, Gene2980, Gene3851, Gene4143 </span></span>
<span><span class="co">## Negative:  Gene4168, Gene716, Gene1947, Gene4044, Gene920, Gene550, Gene266, Gene2145, Gene1540, Gene4026 </span></span>
<span><span class="co">##     Gene4717, Gene4097, Gene303, Gene3369, Gene3528, Gene273, Gene3075, Gene1476, Gene3361, Gene1049 </span></span>
<span><span class="co">##     Gene4700, Gene2991, Gene310, Gene4101, Gene307, Gene3739, Gene4720, Gene268, Gene797, Gene2748 </span></span>
<span><span class="co">## PC_ 2 </span></span>
<span><span class="co">## Positive:  Gene1147, Gene4978, Gene4045, Gene3516, Gene2718, Gene449, Gene4265, Gene1735, Gene2672, Gene4324 </span></span>
<span><span class="co">##     Gene2713, Gene3520, Gene921, Gene807, Gene761, Gene2933, Gene141, Gene4864, Gene1756, Gene1878 </span></span>
<span><span class="co">##     Gene3461, Gene1829, Gene2879, Gene3232, Gene3671, Gene3658, Gene381, Gene1859, Gene2435, Gene75 </span></span>
<span><span class="co">## Negative:  Gene844, Gene4871, Gene3069, Gene487, Gene2671, Gene3600, Gene3683, Gene772, Gene4390, Gene33 </span></span>
<span><span class="co">##     Gene825, Gene2072, Gene1494, Gene1699, Gene3854, Gene4601, Gene3663, Gene2104, Gene4879, Gene2541 </span></span>
<span><span class="co">##     Gene3345, Gene3387, Gene4256, Gene3386, Gene3639, Gene4237, Gene3337, Gene2033, Gene3430, Gene3325 </span></span>
<span><span class="co">## PC_ 3 </span></span>
<span><span class="co">## Positive:  Gene4379, Gene785, Gene3409, Gene1985, Gene3609, Gene3527, Gene1888, Gene3256, Gene1862, Gene2306 </span></span>
<span><span class="co">##     Gene1415, Gene498, Gene3431, Gene194, Gene3945, Gene1439, Gene352, Gene2568, Gene901, Gene2068 </span></span>
<span><span class="co">##     Gene3309, Gene1503, Gene3510, Gene3026, Gene1173, Gene4609, Gene2498, Gene1977, Gene862, Gene1272 </span></span>
<span><span class="co">## Negative:  Gene3375, Gene510, Gene3543, Gene4623, Gene512, Gene2878, Gene224, Gene2465, Gene158, Gene3732 </span></span>
<span><span class="co">##     Gene1732, Gene1331, Gene4316, Gene4530, Gene2909, Gene880, Gene4872, Gene3762, Gene1567, Gene4244 </span></span>
<span><span class="co">##     Gene3298, Gene2382, Gene3000, Gene54, Gene196, Gene683, Gene1090, Gene1159, Gene3965, Gene4326 </span></span>
<span><span class="co">## PC_ 4 </span></span>
<span><span class="co">## Positive:  Gene2249, Gene3443, Gene3527, Gene71, Gene2586, Gene1167, Gene3670, Gene368, Gene3594, Gene4908 </span></span>
<span><span class="co">##     Gene3933, Gene1762, Gene842, Gene4562, Gene394, Gene2133, Gene2230, Gene3718, Gene876, Gene1452 </span></span>
<span><span class="co">##     Gene4025, Gene114, Gene1616, Gene2547, Gene3955, Gene2619, Gene2533, Gene206, Gene3890, Gene575 </span></span>
<span><span class="co">## Negative:  Gene3900, Gene650, Gene854, Gene1761, Gene4723, Gene477, Gene2307, Gene3566, Gene2816, Gene3031 </span></span>
<span><span class="co">##     Gene3629, Gene3551, Gene4966, Gene4342, Gene2144, Gene4904, Gene4368, Gene1026, Gene2907, Gene3115 </span></span>
<span><span class="co">##     Gene2223, Gene2350, Gene1295, Gene960, Gene3112, Gene3069, Gene2855, Gene2780, Gene4066, Gene3024 </span></span>
<span><span class="co">## PC_ 5 </span></span>
<span><span class="co">## Positive:  Gene4030, Gene3767, Gene1452, Gene2575, Gene4562, Gene3762, Gene3185, Gene687, Gene3069, Gene2444 </span></span>
<span><span class="co">##     Gene3157, Gene4018, Gene4784, Gene2569, Gene2685, Gene4368, Gene3031, Gene3040, Gene3115, Gene2149 </span></span>
<span><span class="co">##     Gene4860, Gene2907, Gene3817, Gene671, Gene2405, Gene3442, Gene591, Gene4010, Gene1690, Gene3229 </span></span>
<span><span class="co">## Negative:  Gene394, Gene898, Gene650, Gene447, Gene3050, Gene3908, Gene1167, Gene2159, Gene4543, Gene3527 </span></span>
<span><span class="co">##     Gene93, Gene4557, Gene71, Gene2803, Gene2162, Gene1828, Gene2586, Gene3365, Gene3609, Gene857 </span></span>
<span><span class="co">##     Gene3933, Gene4025, Gene4744, Gene1482, Gene799, Gene4711, Gene3176, Gene842, Gene875, Gene3201</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">se_unintegrated</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">se_unintegrated</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-1-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="run-anglemania">run anglemania<a class="anchor" aria-label="anchor" href="#run-anglemania"></a>
</h2>
<div class="section level3">
<h3 id="create-anglemania-object">create anglemania object<a class="anchor" aria-label="anchor" href="#create-anglemania-object"></a>
</h3>
<ul>
<li>We create an anglemania object to have a standardized input to the
anglemania function
<ul>
<li>It can be created from:
<ul>
<li>An SingleCellExperiment (SCE) object</li>
<li>Seurat object</li>
<li>a list of SCE and/or Seurat objects</li>
</ul>
</li>
</ul>
</li>
<li>pre-requisites are that you:
<ul>
<li>have a column in the metadata (colData for SCE) that specifies the
batch/sample a cell belongs to</li>
<li>or if you pass a list of SCE/Seurat objects, just provide this as
the input and don’t specify a batch key</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">se</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          orig.ident nCount_RNA nFeature_RNA  Cell  Batch  Group ExpLibSize</span></span>
<span><span class="co">## Cell1 SeuratProject      47324         3707 Cell1 Batch1 Group1   46898.14</span></span>
<span><span class="co">## Cell2 SeuratProject      54908         3850 Cell2 Batch1 Group1   54688.18</span></span>
<span><span class="co">## Cell3 SeuratProject      51501         3799 Cell3 Batch1 Group2   52027.93</span></span>
<span><span class="co">## Cell4 SeuratProject      52272         3862 Cell4 Batch1 Group1   52319.47</span></span>
<span><span class="co">## Cell5 SeuratProject      37808         3569 Cell5 Batch1 Group3   37774.72</span></span>
<span><span class="co">## Cell6 SeuratProject      58801         3843 Cell6 Batch1 Group3   58112.35</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch_key</span> <span class="op">&lt;-</span> <span class="st">"Batch"</span></span>
<span></span>
<span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_anglemania_object.html">create_anglemania_object</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>  batch_key <span class="op">=</span> <span class="va">batch_key</span>,</span>
<span>  min_cells_per_gene <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seurat_assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## No dataset_key specified.</span></span>
<span><span class="co">## Assuming that all samples belong to the same dataset and are separated by batch_key: Batch</span></span></code></pre>
<pre><code><span><span class="co">## Using the intersection of filtered genes from all batches...</span></span></code></pre>
<pre><code><span><span class="co">## Number of genes in intersected set: 4964</span></span></code></pre>
<pre><code><span><span class="co">## Extracting count matrices...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering each batch to at least 1 cells per gene...</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span></span></code></pre></div>
<pre><code><span><span class="co">## anglemania_object</span></span>
<span><span class="co">## --------------</span></span>
<span><span class="co">## Dataset key: NA </span></span>
<span><span class="co">## Batch key: Batch </span></span>
<span><span class="co">## Number of datasets: 1 </span></span>
<span><span class="co">## Total number of batches: 4 </span></span>
<span><span class="co">## Batches (showing first 5):</span></span>
<span><span class="co">## Batch1, Batch2, Batch3, Batch4 </span></span>
<span><span class="co">## Number of intersected genes: 4964 </span></span>
<span><span class="co">## Intersected genes (showing first 10):</span></span>
<span><span class="co">## Gene1, Gene2, Gene3, Gene4, Gene5, Gene6, Gene7, Gene8, Gene9, Gene10 , ...</span></span>
<span><span class="co">## Min cells per gene: 1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-anglemania-function">run anglemania function<a class="anchor" aria-label="anchor" href="#run-anglemania-function"></a>
</h3>
<ul>
<li>the anglemania function has five arguments
<ul>
<li>the input anglemania object</li>
<li>method: either cosine, spearman or diem
<ul>
<li>this is the method by which the relationship of the gene pairs is
measured. Default is cosine,</li>
</ul>
</li>
<li>zscore_mean_threshold: We compute a mean of the zscore of the
relationship between a gene pair, and then we set a minimal cutoff for
the (absolute) mean zscore. A cutoff of 2 means that the filtered gene
pairs have a relationship, e.g. cosine similarity, that is 2 standard
deviations away from the mean of all cosine similarities from this
dataset. A higher value means a more stringent cutoff.</li>
<li>zscore_sn_threshold: The SN or signal-to-noise ratio measures the
invariance of the relationship of the relationship between the gene
pair. A high SN ratio means that the relationship is constant over
multiple batches.</li>
<li>max_n_genes: you can specify a maximum number of extracted genes.
They are sorted by decreasing mean zscore</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">angl</span>,</span>
<span>  zscore_mean_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  zscore_sn_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  max_n_genes <span class="op">=</span> <span class="fl">2000</span> <span class="co"># optionally define a max number of genes.</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing angles and transforming to z-scores...</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmps32hWY/file1edd64627473" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmps32hWY/file1edd56e4089" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmps32hWY/file1edd29fc883f" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Creating directory "/tmp/Rtmps32hWY/file1edde052a84" which didn't exist..</span></span></code></pre>
<pre><code><span><span class="co">## Computing statistics...</span></span></code></pre>
<pre><code><span><span class="co">## Weighting matrix_list...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating mean...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating sds...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering features...</span></span></code></pre>
<pre><code><span><span class="co">## Selecting features...</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Selected 1496 genes for integration."</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemanise_utils.html">select_genes</a></span><span class="op">(</span><span class="va">angl</span>,</span>
<span>                     zscore_mean_threshold <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>                     zscore_sn_threshold <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Selected 306 genes for integration."</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect the anglemania genes</span></span>
<span><span class="va">integration_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_object-methods.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">integration_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Gene394"  "Gene3527" "Gene71"   "Gene5000" "Gene1055" "Gene4201"</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">integration_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 306</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="integration">integration<a class="anchor" aria-label="anchor" href="#integration"></a>
</h3>
<ul>
<li>we implemented the easy-to-use integrate_by_features() function from
the anglemania package which uses Seurat CCA integration</li>
<li>you can also just use the anglemania genes for other integration
algorithms</li>
</ul>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if the Seurat FindIntegrationAnchors() function does not work,</span></span>
<span><span class="co"># change this to the specified size:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">4000</span> <span class="op">*</span> <span class="fl">1024</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">seurat_integrated_angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/integrate_by_features.html">integrate_by_features</a></span><span class="op">(</span><span class="va">se</span>,</span>
<span>    <span class="va">angl</span>,</span>
<span>    process <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Log normalizing data...</span></span></code></pre>
<pre><code><span><span class="co">## Finding integration anchors...</span></span></code></pre>
<pre><code><span><span class="co">## Integrating samples...</span></span></code></pre>
<pre><code><span><span class="co">## Running PCA with 30 PCs</span></span></code></pre>
<pre><code><span><span class="co">## Running UMAP with 30 PCs and 30 neighbors</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_integrated_angl</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 5306 features across 3600 samples within 2 assays </span></span>
<span><span class="co">## Active assay: integrated (306 features, 306 variable features)</span></span>
<span><span class="co">##  2 layers present: data, scale.data</span></span>
<span><span class="co">##  1 other assay present: RNA</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, umap</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="plot-integration">plot integration<a class="anchor" aria-label="anchor" href="#plot-integration"></a>
</h3>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated_angl</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated_angl</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-2-2.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="comparison-to-using-hvgs">Comparison to using HVGs<a class="anchor" aria-label="anchor" href="#comparison-to-using-hvgs"></a>
</h2>
<div class="section level3">
<h3 id="using-hvgs">using HVGs<a class="anchor" aria-label="anchor" href="#using-hvgs"></a>
</h3>
<div class="section level4">
<h4 id="get-hvgs">get HVGs<a class="anchor" aria-label="anchor" href="#get-hvgs"></a>
</h4>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SplitObject.html" class="external-link">SplitObject</a></span><span class="op">(</span><span class="va">se</span>, split.by <span class="op">=</span> <span class="va">batch_key</span><span class="op">)</span></span>
<span><span class="va">se_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">se_list</span>, <span class="va">NormalizeData</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Normalizing layer: counts</span></span>
<span><span class="co">## Normalizing layer: counts</span></span>
<span><span class="co">## Normalizing layer: counts</span></span>
<span><span class="co">## Normalizing layer: counts</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># anglemania chose only 324 genes based on the thresholds</span></span>
<span><span class="co">#   to make it comparable we only choose 324 highly variable genes:</span></span>
<span><span class="va">hvg_features</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SelectIntegrationFeatures.html" class="external-link">SelectIntegrationFeatures</a></span><span class="op">(</span><span class="va">se_list</span>, nfeatures <span class="op">=</span> <span class="fl">324</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## No variable features found for object1 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<pre><code><span><span class="co">## No variable features found for object2 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<pre><code><span><span class="co">## No variable features found for object3 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<pre><code><span><span class="co">## No variable features found for object4 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="integration-1">integration<a class="anchor" aria-label="anchor" href="#integration-1"></a>
</h4>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">seurat_integrated_hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/integrate_seurat_list.html">integrate_seurat_list</a></span><span class="op">(</span><span class="va">se_list</span>,</span>
<span>    features <span class="op">=</span> <span class="va">hvg_features</span>,</span>
<span>    process <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Log normalizing data...</span></span></code></pre>
<pre><code><span><span class="co">## Finding integration anchors...</span></span></code></pre>
<pre><code><span><span class="co">## Integrating samples...</span></span></code></pre>
<pre><code><span><span class="co">## Running PCA with 30 PCs</span></span></code></pre>
<pre><code><span><span class="co">## Running UMAP with 30 PCs and 30 neighbors</span></span></code></pre>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_integrated_hvg</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 5324 features across 3600 samples within 2 assays </span></span>
<span><span class="co">## Active assay: integrated (324 features, 324 variable features)</span></span>
<span><span class="co">##  2 layers present: data, scale.data</span></span>
<span><span class="co">##  1 other assay present: RNA</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, umap</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="plot-integration-1">plot integration<a class="anchor" aria-label="anchor" href="#plot-integration-1"></a>
</h4>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated_hvg</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated_hvg</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-3-2.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="using-2000-hvgs">using 2000 HVGs<a class="anchor" aria-label="anchor" href="#using-2000-hvgs"></a>
</h3>
<div class="section level4">
<h4 id="get-hvgs-1">get HVGs<a class="anchor" aria-label="anchor" href="#get-hvgs-1"></a>
</h4>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hvg_features</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SelectIntegrationFeatures.html" class="external-link">SelectIntegrationFeatures</a></span><span class="op">(</span><span class="va">se_list</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## No variable features found for object1 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<pre><code><span><span class="co">## No variable features found for object2 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<pre><code><span><span class="co">## No variable features found for object3 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<pre><code><span><span class="co">## No variable features found for object4 in the object.list. Running FindVariableFeatures ...</span></span></code></pre>
<pre><code><span><span class="co">## Finding variable features for layer counts</span></span></code></pre>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="va">seurat_integrated_hvg_2000</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/integrate_seurat_list.html">integrate_seurat_list</a></span><span class="op">(</span><span class="va">se_list</span>,</span>
<span>    features <span class="op">=</span> <span class="va">hvg_features</span>,</span>
<span>    process <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Log normalizing data...</span></span></code></pre>
<pre><code><span><span class="co">## Finding integration anchors...</span></span></code></pre>
<pre><code><span><span class="co">## Integrating samples...</span></span></code></pre>
<pre><code><span><span class="co">## Running PCA with 30 PCs</span></span></code></pre>
<pre><code><span><span class="co">## Running UMAP with 30 PCs and 30 neighbors</span></span></code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_integrated_hvg_2000</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 7000 features across 3600 samples within 2 assays </span></span>
<span><span class="co">## Active assay: integrated (2000 features, 2000 variable features)</span></span>
<span><span class="co">##  2 layers present: data, scale.data</span></span>
<span><span class="co">##  1 other assay present: RNA</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, umap</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="plot-integration-2">plot integration<a class="anchor" aria-label="anchor" href="#plot-integration-2"></a>
</h4>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated_hvg_2000</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seurat_integrated_hvg_2000</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="anglemania_tutorial_Seurat_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] future_1.40.0               splatter_1.31.1            </span></span>
<span><span class="co">##  [3] SingleCellExperiment_1.29.2 SummarizedExperiment_1.37.0</span></span>
<span><span class="co">##  [5] Biobase_2.67.0              GenomicRanges_1.59.1       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.43.4         IRanges_2.41.3             </span></span>
<span><span class="co">##  [9] S4Vectors_0.45.4            BiocGenerics_0.53.6        </span></span>
<span><span class="co">## [11] generics_0.1.3              MatrixGenerics_1.19.1      </span></span>
<span><span class="co">## [13] matrixStats_1.5.0           Seurat_5.2.1               </span></span>
<span><span class="co">## [15] SeuratObject_5.0.2          sp_2.2-0                   </span></span>
<span><span class="co">## [17] dplyr_1.1.4                 anglemania_0.99.1          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RcppAnnoy_0.0.22        splines_4.5.0           later_1.4.2            </span></span>
<span><span class="co">##   [4] tibble_3.2.1            polyclip_1.10-7         fastDummies_1.7.5      </span></span>
<span><span class="co">##   [7] lifecycle_1.0.4         doParallel_1.0.17       globals_0.16.3         </span></span>
<span><span class="co">##  [10] lattice_0.22-6          MASS_7.3-65             backports_1.5.0        </span></span>
<span><span class="co">##  [13] magrittr_2.0.3          plotly_4.10.4           sass_0.4.10            </span></span>
<span><span class="co">##  [16] rmarkdown_2.29          jquerylib_0.1.4         yaml_2.3.10            </span></span>
<span><span class="co">##  [19] bigparallelr_0.3.2      httpuv_1.6.15           sctransform_0.4.1      </span></span>
<span><span class="co">##  [22] spam_2.11-1             spatstat.sparse_3.1-0   reticulate_1.42.0      </span></span>
<span><span class="co">##  [25] cowplot_1.1.3           pbapply_1.7-2           RColorBrewer_1.1-3     </span></span>
<span><span class="co">##  [28] abind_1.4-8             Rtsne_0.17              purrr_1.0.4            </span></span>
<span><span class="co">##  [31] bigassertr_0.1.6        GenomeInfoDbData_1.2.14 ggrepel_0.9.6          </span></span>
<span><span class="co">##  [34] irlba_2.3.5.1           listenv_0.9.1           spatstat.utils_3.1-3   </span></span>
<span><span class="co">##  [37] goftest_1.2-3           RSpectra_0.16-2         spatstat.random_3.3-3  </span></span>
<span><span class="co">##  [40] fitdistrplus_1.2-2      parallelly_1.43.0       pkgdown_2.1.1          </span></span>
<span><span class="co">##  [43] codetools_0.2-20        DelayedArray_0.33.6     tidyselect_1.2.1       </span></span>
<span><span class="co">##  [46] UCSC.utils_1.3.1        farver_2.1.2            bigstatsr_1.6.1        </span></span>
<span><span class="co">##  [49] spatstat.explore_3.4-2  flock_0.7               jsonlite_2.0.0         </span></span>
<span><span class="co">##  [52] progressr_0.15.1        ggridges_0.5.6          survival_3.8-3         </span></span>
<span><span class="co">##  [55] iterators_1.0.14        systemfonts_1.2.2       foreach_1.5.2          </span></span>
<span><span class="co">##  [58] tools_4.5.0             ragg_1.4.0              ica_1.0-3              </span></span>
<span><span class="co">##  [61] Rcpp_1.0.14             glue_1.8.0              gridExtra_2.3          </span></span>
<span><span class="co">##  [64] SparseArray_1.7.7       xfun_0.52               withr_3.0.2            </span></span>
<span><span class="co">##  [67] fastmap_1.2.0           digest_0.6.37           R6_2.6.1               </span></span>
<span><span class="co">##  [70] mime_0.13               textshaping_1.0.0       colorspace_2.1-1       </span></span>
<span><span class="co">##  [73] scattermore_1.2         tensor_1.5              spatstat.data_3.1-6    </span></span>
<span><span class="co">##  [76] tidyr_1.3.1             data.table_1.17.0       httr_1.4.7             </span></span>
<span><span class="co">##  [79] htmlwidgets_1.6.4       S4Arrays_1.7.3          uwot_0.2.3             </span></span>
<span><span class="co">##  [82] pkgconfig_2.0.3         gtable_0.3.6            lmtest_0.9-40          </span></span>
<span><span class="co">##  [85] XVector_0.47.2          htmltools_0.5.8.1       dotCall64_1.2          </span></span>
<span><span class="co">##  [88] scales_1.3.0            png_0.1-8               spatstat.univar_3.1-2  </span></span>
<span><span class="co">##  [91] knitr_1.50              reshape2_1.4.4          checkmate_2.3.2        </span></span>
<span><span class="co">##  [94] nlme_3.1-168            cachem_1.1.0            zoo_1.8-14             </span></span>
<span><span class="co">##  [97] stringr_1.5.1           rmio_0.4.0              KernSmooth_2.23-26     </span></span>
<span><span class="co">## [100] parallel_4.5.0          miniUI_0.1.1.1          desc_1.4.3             </span></span>
<span><span class="co">## [103] pillar_1.10.2           grid_4.5.0              vctrs_0.6.5            </span></span>
<span><span class="co">## [106] RANN_2.6.2              promises_1.3.2          ff_4.5.2               </span></span>
<span><span class="co">## [109] xtable_1.8-4            cluster_2.1.8.1         evaluate_1.0.3         </span></span>
<span><span class="co">## [112] cli_3.6.4               locfit_1.5-9.12         compiler_4.5.0         </span></span>
<span><span class="co">## [115] rlang_1.1.6             crayon_1.5.3            future.apply_1.11.3    </span></span>
<span><span class="co">## [118] labeling_0.4.3          ps_1.9.1                plyr_1.8.9             </span></span>
<span><span class="co">## [121] fs_1.6.6                stringi_1.8.7           viridisLite_0.4.2      </span></span>
<span><span class="co">## [124] deldir_2.0-4            BiocParallel_1.41.5     munsell_0.5.1          </span></span>
<span><span class="co">## [127] lazyeval_0.2.2          spatstat.geom_3.3-6     Matrix_1.7-3           </span></span>
<span><span class="co">## [130] RcppHNSW_0.6.0          patchwork_1.3.0         ggplot2_3.5.2          </span></span>
<span><span class="co">## [133] shiny_1.10.0            ROCR_1.0-11             igraph_2.1.4           </span></span>
<span><span class="co">## [136] bslib_0.9.0             bit_4.6.0</span></span></code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Aaron Kollotzek, Vedran Franke, Artem Baranovskii, Altuna Akalin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
