<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Functions_Showcasing â€¢ anglemania</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Functions_Showcasing">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">anglemania</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/anglemania_tutorial_Seurat.html">Running anglemania with Seurat</a>
    </li>
    <li>
      <a href="../articles/anglemania_tutorial.html">Running anglemania with SingleCellExperiment</a>
    </li>
    <li>
      <a href="../articles/Functions_Showcasing.html">Functions_Showcasing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMSBbioinfo/anglemania/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Functions_Showcasing</h1>
            
            <h4 data-toc-skip class="date">2025-03-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BIMSBbioinfo/anglemania/blob/master/vignettes/Functions_Showcasing.Rmd" class="external-link"><code>vignettes/Functions_Showcasing.Rmd</code></a></small>
      <div class="hidden name"><code>Functions_Showcasing.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this Vignette we want to quickly show the few other functions that
are available in anglemania besides the two main ones:
<code>create_anglemania_object</code> and <code>anglemania</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinformatics.mdc-berlin.de/anglemania" class="external-link">anglemania</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SummarizedExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="normal-anglemania-workflow">Normal anglemania workflow<a class="anchor" aria-label="anchor" href="#normal-anglemania-workflow"></a>
</h2>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">batch_key</span> <span class="op">&lt;-</span> <span class="st">"batch"</span></span>
<span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_anglemania_object.html">create_anglemania_object</a></span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="va">batch_key</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## No dataset_key specified.</span></span>
<span><span class="co">## Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span></span></code></pre>
<pre><code><span><span class="co">## Extracting count matrices...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering each batch to at least 1 cells per gene...</span></span></code></pre>
<pre><code><span><span class="co">## Using the intersection of filtered genes from all batches...</span></span></code></pre>
<pre><code><span><span class="co">## Number of genes in intersected set: 300</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing angles and transforming to z-scores...</span></span></code></pre>
<pre><code><span><span class="co">## Computing statistics...</span></span></code></pre>
<pre><code><span><span class="co">## Weighting matrix_list...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating mean...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating sds...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering features...</span></span></code></pre>
<pre><code><span><span class="co">## Selecting features...</span></span></code></pre>
<pre><code><span><span class="co">## If desired, you can re-run the selection of genes with a lower zscore_mean_threshold and/or zscore_sn_threshold by using the 'select_genes' function. e.g.: angl &lt;- select_genes(angl, zscore_mean_threshold = 1, zscore_sn_threshold = 1, max_n_genes = 2000)</span></span></code></pre>
<pre><code><span><span class="co">## Please inspect get_anglemania_genes(angl)$info for info on the scores of the selected gene pairs.</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Selected 36 genes for integration."</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span></span></code></pre></div>
<pre><code><span><span class="co">## anglemania_object</span></span>
<span><span class="co">## --------------</span></span>
<span><span class="co">## Dataset key: NA </span></span>
<span><span class="co">## Batch key: batch </span></span>
<span><span class="co">## Number of datasets: 1 </span></span>
<span><span class="co">## Total number of batches: 2 </span></span>
<span><span class="co">## Batches (showing first 5):</span></span>
<span><span class="co">## batch1, batch2 </span></span>
<span><span class="co">## Number of intersected genes: 300 </span></span>
<span><span class="co">## Intersected genes (showing first 10):</span></span>
<span><span class="co">## gene1, gene2, gene3, gene4, gene5, gene6, gene7, gene8, gene9, gene10 , ...</span></span>
<span><span class="co">## Min cells per gene: 1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="under-the-hood-of-the-anglemania-function">under the hood of the anglemania function<a class="anchor" aria-label="anchor" href="#under-the-hood-of-the-anglemania-function"></a>
</h2>
<ul>
<li>anglemania is run on the anglemania object and it calls three
functions:
<ul>
<li>factorise:
<ul>
<li>creates a permutation of the input matrix to create a null
distribution</li>
<li>computes the cosine similarity (or spearman coefficient or diem)
between gene expresion vector pairs matrix for both the original and
permuted matrices</li>
<li>computes the zscore of the relationship between the gene pairs
taking the mean and standard deviation of the null distribution</li>
<li>it does this for every batch in the dataset!</li>
</ul>
</li>
<li>get_list_stats
<ul>
<li>computes the mean and standard deviation of the zscores across the
matrices from the different batches.</li>
</ul>
</li>
<li>select_genes
<ul>
<li>filters the gene pairs by the mean zscore and the signal-to-noise
ratio (SN, i.e.Â the mean divided by the standard deviation).</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="section level3">
<h3 id="factorise">factorise<a class="anchor" aria-label="anchor" href="#factorise"></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barcodes_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[[</span><span class="va">batch_key</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">counts_by_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">barcodes_by_batch</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">[</span>, <span class="va">x</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/sparse_to_fbm.html">sparse_to_fbm</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">counts_by_batch</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##       [,1] [,2] [,3] [,4] [,5] [,6]</span></span>
<span><span class="co">##  [1,]    8    5    5    2    4    2</span></span>
<span><span class="co">##  [2,]    9    5    4    1    4    9</span></span>
<span><span class="co">##  [3,]    4    2    7    4    6    1</span></span>
<span><span class="co">##  [4,]    7    4    4    1    3    8</span></span>
<span><span class="co">##  [5,]    6    8    5    5    7    4</span></span>
<span><span class="co">##  [6,]    5    8    5    9    8    6</span></span>
<span><span class="co">##  [7,]    6    4    7    4    7    5</span></span>
<span><span class="co">##  [8,]    3    3    3    3    5    4</span></span>
<span><span class="co">##  [9,]    6    4    5    5    3    1</span></span>
<span><span class="co">## [10,]    6    4    5    2    6    4</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">counts_by_batch</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "FBM"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "bigstatsr"</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factorised</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">counts_by_batch</span>, <span class="va">factorise</span><span class="op">)</span></span>
<span><span class="va">factorised</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##             [,1]        [,2]       [,3]        [,4]       [,5]       [,6]</span></span>
<span><span class="co">##  [1,]         NA -1.77559431 -0.8253279  1.14791468  0.8304337  0.4723581</span></span>
<span><span class="co">##  [2,] -1.7755943          NA  0.5726423  2.49881313  0.5105572  1.4157996</span></span>
<span><span class="co">##  [3,] -0.8253279  0.57264234         NA  0.32316488 -1.5416407 -0.0291761</span></span>
<span><span class="co">##  [4,]  1.1479147  2.49881313  0.3231649          NA -0.3972312 -0.6038679</span></span>
<span><span class="co">##  [5,]  0.8304337  0.51055725 -1.5416407 -0.39723123         NA -0.2773380</span></span>
<span><span class="co">##  [6,]  0.4723581  1.41579962 -0.0291761 -0.60386791 -0.2773380         NA</span></span>
<span><span class="co">##  [7,] -0.3793546 -0.08986951 -1.0834393  0.01453768  0.3333673  1.3412298</span></span>
<span><span class="co">##  [8,]  0.0815372 -1.42754363 -1.3398126 -0.27740608 -0.8317583 -1.6147436</span></span>
<span><span class="co">##  [9,]  0.5330219  0.05725224 -0.3868227 -0.30696466 -0.9537188 -0.1300575</span></span>
<span><span class="co">## [10,]  0.1071082 -0.33100426  1.7263481  0.43416665  0.5866818  1.1300376</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="get_list_stats">get_list_stats<a class="anchor" aria-label="anchor" href="#get_list_stats"></a>
</h3>
<ul>
<li>get_list_stats gets called on the anglemania object so we create and
inspect a anglemania object</li>
</ul>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sce_example.html">sce_example</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">batch_key</span> <span class="op">&lt;-</span> <span class="st">"batch"</span></span>
<span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_anglemania_object.html">create_anglemania_object</a></span><span class="op">(</span><span class="va">sce</span>, batch_key <span class="op">=</span> <span class="va">batch_key</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## No dataset_key specified.</span></span>
<span><span class="co">## Assuming that all samples belong to the same dataset and are separated by batch_key: batch</span></span></code></pre>
<pre><code><span><span class="co">## Extracting count matrices...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering each batch to at least 1 cells per gene...</span></span></code></pre>
<pre><code><span><span class="co">## Using the intersection of filtered genes from all batches...</span></span></code></pre>
<pre><code><span><span class="co">## Number of genes in intersected set: 300</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/anglemania_object-methods.html">list_stats</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span> <span class="co"># this slot is empty</span></span></code></pre></div>
<pre><code><span><span class="co">## list()</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania.html">anglemania</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Computing angles and transforming to z-scores...</span></span></code></pre>
<pre><code><span><span class="co">## Computing statistics...</span></span></code></pre>
<pre><code><span><span class="co">## Weighting matrix_list...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating mean...</span></span></code></pre>
<pre><code><span><span class="co">## Calculating sds...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering features...</span></span></code></pre>
<pre><code><span><span class="co">## Selecting features...</span></span></code></pre>
<pre><code><span><span class="co">## If desired, you can re-run the selection of genes with a lower zscore_mean_threshold and/or zscore_sn_threshold by using the 'select_genes' function. e.g.: angl &lt;- select_genes(angl, zscore_mean_threshold = 1, zscore_sn_threshold = 1, max_n_genes = 2000)</span></span></code></pre>
<pre><code><span><span class="co">## Please inspect get_anglemania_genes(angl)$info for info on the scores of the selected gene pairs.</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Selected 36 genes for integration."</span></span></code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="../reference/anglemania_object-methods.html">list_stats</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "mean_zscore" "sds_zscore"  "sn_zscore"   "prefiltered"</span></span></code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="fu"><a href="../reference/anglemania_object-methods.html">list_stats</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/anglemania_object-methods.html">list_stats</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span><span class="op">$</span><span class="va">mean_zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##             [,1]        [,2]       [,3]        [,4]        [,5]        [,6]</span></span>
<span><span class="co">##  [1,]         NA -1.47361566 -0.6737199  0.79226342  0.20563482  0.29347030</span></span>
<span><span class="co">##  [2,] -1.4736157          NA  0.7911388  1.78368051  0.85782278  0.05812042</span></span>
<span><span class="co">##  [3,] -0.6737199  0.79113876         NA -0.81497507 -0.16581743  0.40981298</span></span>
<span><span class="co">##  [4,]  0.7922634  1.78368051 -0.8149751          NA -0.32367957 -0.32209876</span></span>
<span><span class="co">##  [5,]  0.2056348  0.85782278 -0.1658174 -0.32367957          NA -0.82414755</span></span>
<span><span class="co">##  [6,]  0.2934703  0.05812042  0.4098130 -0.32209876 -0.82414755          NA</span></span>
<span><span class="co">##  [7,]  0.5353808  0.39957517 -1.3597171 -0.05621663  0.31106411  0.48942810</span></span>
<span><span class="co">##  [8,] -0.0478384 -1.11509607 -0.7896121 -0.63001026 -1.43434185 -0.84468556</span></span>
<span><span class="co">##  [9,]  1.2866480  0.34648068 -0.5253048 -0.78855628 -0.69594257 -0.50281463</span></span>
<span><span class="co">## [10,]  0.5280248  1.21974236  0.8667575 -0.27678173  0.04317401  0.87613079</span></span></code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/anglemania_object-methods.html">list_stats</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span><span class="op">$</span><span class="va">sn_zscore</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##            [,1]       [,2]      [,3]      [,4]        [,5]       [,6]</span></span>
<span><span class="co">##  [1,]        NA 4.87986702 4.4438254 2.2276412  0.32912160 1.64052702</span></span>
<span><span class="co">##  [2,] 4.8798670         NA 3.6208316 2.4941954  2.47022152 0.04280865</span></span>
<span><span class="co">##  [3,] 4.4438254 3.62083164        NA 0.7160588  0.12052233 0.93353798</span></span>
<span><span class="co">##  [4,] 2.2276412 2.49419544 0.7160588        NA  4.40071059 1.14312995</span></span>
<span><span class="co">##  [5,] 0.3291216 2.47022152 0.1205223 4.4007106          NA 1.50719307</span></span>
<span><span class="co">##  [6,] 1.6405270 0.04280865 0.9335380 1.1431300  1.50719307         NA</span></span>
<span><span class="co">##  [7,] 0.5852849 0.81638473 4.9215576 0.7945330 13.94709308 0.57457988</span></span>
<span><span class="co">##  [8,] 0.3697637 3.56890627 1.4351352 1.7867351  2.38032046 1.09691151</span></span>
<span><span class="co">##  [9,] 1.7072764 1.19794817 3.7933035 1.6373962  2.69979335 1.34890662</span></span>
<span><span class="co">## [10,] 1.2544641 0.78655168 1.0083377 0.3893134  0.07943585 3.45060039</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="select_genes">select_genes<a class="anchor" aria-label="anchor" href="#select_genes"></a>
</h3>
<ul>
<li>under the hood, anglemania calls select_genes on the anglemania
object with the default thresholds zscore_mean_threshold = 2.5,
zscore_sn_threshold = 2.5</li>
<li>we can use select_genes to change the thresholds without having to
run anglemania again</li>
</ul>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">previous_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_object-methods.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span>
<span><span class="va">angl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_genes.html">select_genes</a></span><span class="op">(</span><span class="va">angl</span>,</span>
<span>                     zscore_mean_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                     zscore_sn_threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## If desired, you can re-run the selection of genes with a lower zscore_mean_threshold and/or zscore_sn_threshold by using the 'select_genes' function. e.g.: angl &lt;- select_genes(angl, zscore_mean_threshold = 1, zscore_sn_threshold = 1, max_n_genes = 2000)</span></span></code></pre>
<pre><code><span><span class="co">## Please inspect get_anglemania_genes(angl)$info for info on the scores of the selected gene pairs.</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Selected 214 genes for integration."</span></span></code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inspect the anglemania genes</span></span>
<span><span class="va">new_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anglemania_object-methods.html">get_anglemania_genes</a></span><span class="op">(</span><span class="va">angl</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">previous_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 36</span></span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">new_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 214</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Aaron Kollotzek, Vedran Franke, Artem Baranovskii, Altuna Akalin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
